--- a/net/minecraft/entity/projectile/FishingBobberEntity.java
+++ b/net/minecraft/entity/projectile/FishingBobberEntity.java
@@ -210,8 +210,8 @@
    private boolean removeIfInvalid() {
       ItemStack itemstack = this.owner.getMainHandStack();
       ItemStack itemstack1 = this.owner.getOffHandStack();
-      boolean flag = itemstack.getItem() == Items.FISHING_ROD;
-      boolean flag1 = itemstack1.getItem() == Items.FISHING_ROD;
+      boolean flag = itemstack.getItem() instanceof net.minecraft.item.FishingRodItem;
+      boolean flag1 = itemstack1.getItem() instanceof net.minecraft.item.FishingRodItem;
       if (!this.owner.removed && this.owner.isAlive() && (flag || flag1) && !(this.squaredDistanceTo(this.owner) > 1024.0D)) {
          return false;
       } else {
@@ -295,7 +295,7 @@
             double d1 = (double)((float)MathHelper.floor(this.getY()) + 1.0F);
             double d2 = this.getZ() + (double)(f2 * (float)this.fishTravelCountdown * 0.1F);
             Block block = serverworld.getBlockState(new BlockPos(d0, d1 - 1.0D, d2)).getBlock();
-            if (block == Blocks.WATER) {
+            if (serverworld.getBlockState(new BlockPos((int)d0, (int)d1 - 1, (int)d2)).getMaterial() == net.minecraft.block.Material.WATER) {
                if (this.random.nextFloat() < 0.15F) {
                   serverworld.spawnParticles(ParticleTypes.BUBBLE, d0, d1 - (double)0.1F, d2, 1, (double)f1, 0.1D, (double)f2, 0.0D);
                }
@@ -332,7 +332,7 @@
             double d5 = (double)((float)MathHelper.floor(this.getY()) + 1.0F);
             double d6 = this.getZ() + (double)(MathHelper.cos(f6) * f7 * 0.1F);
             Block block1 = serverworld.getBlockState(new BlockPos(d4, d5 - 1.0D, d6)).getBlock();
-            if (block1 == Blocks.WATER) {
+            if (serverworld.getBlockState(new BlockPos(d4, d5 - 1.0D, d6)).getMaterial() == net.minecraft.block.Material.WATER) {
                serverworld.spawnParticles(ParticleTypes.SPLASH, d4, d5, d6, 2 + this.random.nextInt(2), (double)0.1F, 0.0D, (double)0.1F, 0.0D);
             }
          }
@@ -357,6 +357,7 @@
    public int use(ItemStack p_146034_1_) {
       if (!this.world.isClient && this.owner != null) {
          int i = 0;
+         net.minecraftforge.event.entity.player.ItemFishedEvent event = null;
          if (this.hookedEntity != null) {
             this.pullHookedEntity();
             Criterions.FISHING_ROD_HOOKED.trigger((ServerPlayerEntity)this.owner, p_146034_1_, this, Collections.emptyList());
@@ -364,8 +365,15 @@
             i = this.hookedEntity instanceof ItemEntity ? 3 : 5;
          } else if (this.hookCountdown > 0) {
             LootContext.Builder lootcontext$builder = (new LootContext.Builder((ServerWorld)this.world)).put(LootContextParameters.POSITION, new BlockPos(this)).put(LootContextParameters.TOOL, p_146034_1_).setRandom(this.random).setLuck((float)this.luckOfTheSeaLevel + this.owner.getLuck());
+            lootcontext$builder.put(LootContextParameters.KILLER_ENTITY, this.owner).put(LootContextParameters.THIS_ENTITY, this);
             LootTable loottable = this.world.getServer().getLootManager().getSupplier(LootTables.FISHING_GAMEPLAY);
             List<ItemStack> list = loottable.getDrops(lootcontext$builder.build(LootContextTypes.FISHING));
+            event = new net.minecraftforge.event.entity.player.ItemFishedEvent(list, this.stuckOnBlock ? 2 : 1, this);
+            net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(event);
+            if (event.isCanceled()) {
+               this.remove();
+               return event.getRodDamage();
+            }
             Criterions.FISHING_ROD_HOOKED.trigger((ServerPlayerEntity)this.owner, p_146034_1_, this, list);
 
             for(ItemStack itemstack : list) {
@@ -390,7 +398,7 @@
          }
 
          this.remove();
-         return i;
+         return event == null ? i : event.getRodDamage();
       } else {
          return 0;
       }
@@ -416,8 +424,9 @@
       return false;
    }
 
-   public void remove() {
-      super.remove();
+   @Override
+   public void remove(boolean keepData) {
+      super.remove(keepData);
       if (this.owner != null) {
          this.owner.fishHook = null;
       }
