--- a/net/minecraft/server/network/ServerHandshakeNetworkHandler.java
+++ b/net/minecraft/server/network/ServerHandshakeNetworkHandler.java
@@ -20,6 +20,7 @@
    }
 
    public void onHandshake(HandshakeC2SPacket packetIn) {
+      if (!net.minecraftforge.fml.server.ServerLifecycleHooks.handleServerLogin(packetIn, this.connection)) return;
       switch(packetIn.getIntendedState()) {
       case LOGIN:
          this.connection.setState(NetworkState.LOGIN);
