--- a/net/minecraft/item/AxeItem.java
+++ b/net/minecraft/item/AxeItem.java
@@ -21,7 +21,7 @@
    protected static final Map<Block, Block> STRIPPED_BLOCKS = (new Builder<Block, Block>()).put(Blocks.OAK_WOOD, Blocks.STRIPPED_OAK_WOOD).put(Blocks.OAK_LOG, Blocks.STRIPPED_OAK_LOG).put(Blocks.DARK_OAK_WOOD, Blocks.STRIPPED_DARK_OAK_WOOD).put(Blocks.DARK_OAK_LOG, Blocks.STRIPPED_DARK_OAK_LOG).put(Blocks.ACACIA_WOOD, Blocks.STRIPPED_ACACIA_WOOD).put(Blocks.ACACIA_LOG, Blocks.STRIPPED_ACACIA_LOG).put(Blocks.BIRCH_WOOD, Blocks.STRIPPED_BIRCH_WOOD).put(Blocks.BIRCH_LOG, Blocks.STRIPPED_BIRCH_LOG).put(Blocks.JUNGLE_WOOD, Blocks.STRIPPED_JUNGLE_WOOD).put(Blocks.JUNGLE_LOG, Blocks.STRIPPED_JUNGLE_LOG).put(Blocks.SPRUCE_WOOD, Blocks.STRIPPED_SPRUCE_WOOD).put(Blocks.SPRUCE_LOG, Blocks.STRIPPED_SPRUCE_LOG).build();
 
    public AxeItem(ToolMaterial tier, float attackDamageIn, float attackSpeedIn, Item.Settings builder) {
-      super(attackDamageIn, attackSpeedIn, tier, EFFECTIVE_BLOCKS, builder);
+      super(attackDamageIn, attackSpeedIn, tier, EFFECTIVE_BLOCKS, builder.addToolType(net.minecraftforge.common.ToolType.AXE, tier.getMiningLevel()));
    }
 
    public float getMiningSpeed(ItemStack stack, BlockState state) {
