--- a/net/minecraft/item/BoneMealItem.java
+++ b/net/minecraft/item/BoneMealItem.java
@@ -29,7 +29,7 @@
       World world = context.getWorld();
       BlockPos blockpos = context.getBlockPos();
       BlockPos blockpos1 = blockpos.offset(context.getSide());
-      if (useOnFertilizable(context.getStack(), world, blockpos)) {
+      if (applyBonemeal(context.getStack(), world, blockpos, context.getPlayer())) {
          if (!world.isClient) {
             world.playLevelEvent(2005, blockpos, 0);
          }
@@ -50,8 +50,17 @@
       }
    }
 
+   @Deprecated //Forge: Use Player/Hand version
    public static boolean useOnFertilizable(ItemStack stack, World worldIn, BlockPos pos) {
+      if (worldIn instanceof net.minecraft.server.world.ServerWorld)
+         return applyBonemeal(stack, worldIn, pos, net.minecraftforge.common.util.FakePlayerFactory.getMinecraft((net.minecraft.server.world.ServerWorld)worldIn));
+      return false;
+   }
+
+   public static boolean applyBonemeal(ItemStack stack, World worldIn, BlockPos pos, net.minecraft.entity.player.PlayerEntity player) {
       BlockState blockstate = worldIn.getBlockState(pos);
+      int hook = net.minecraftforge.event.ForgeEventFactory.onApplyBonemeal(player, worldIn, pos, blockstate, stack);
+      if (hook != 0) return hook > 0;
       if (blockstate.getBlock() instanceof Fertilizable) {
          Fertilizable igrowable = (Fertilizable)blockstate.getBlock();
          if (igrowable.isFertilizable(worldIn, pos, blockstate, worldIn.isClient)) {
@@ -89,7 +98,9 @@
                   }
                }
 
-               if (biome == Biomes.WARM_OCEAN || biome == Biomes.DEEP_WARM_OCEAN) {
+               // FORGE: Use BiomeDictionary here to allow modded warm ocean biomes to spawn coral from bonemeal
+               if (net.minecraftforge.common.BiomeDictionary.hasType(biome, net.minecraftforge.common.BiomeDictionary.Type.OCEAN)
+                       && net.minecraftforge.common.BiomeDictionary.hasType(biome, net.minecraftforge.common.BiomeDictionary.Type.HOT)) {
                   if (i == 0 && side != null && side.getAxis().isHorizontal()) {
                      blockstate = BlockTags.WALL_CORALS.getRandom(worldIn.random).getDefaultState().with(DeadCoralWallFanBlock.FACING, side);
                   } else if (RANDOM.nextInt(4) == 0) {
@@ -128,7 +139,7 @@
       }
 
       BlockState blockstate = worldIn.getBlockState(posIn);
-      if (!blockstate.isAir()) {
+      if (!blockstate.isAir(worldIn, posIn)) {
          for(int i = 0; i < data; ++i) {
             double d0 = RANDOM.nextGaussian() * 0.02D;
             double d1 = RANDOM.nextGaussian() * 0.02D;
