--- a/net/minecraft/item/FishBucketItem.java
+++ b/net/minecraft/item/FishBucketItem.java
@@ -27,11 +27,19 @@
 public class FishBucketItem extends BucketItem {
    private final EntityType<?> fishType;
 
+   @Deprecated
    public FishBucketItem(EntityType<?> fishTypeIn, Fluid p_i49022_2_, Item.Settings builder) {
       super(p_i49022_2_, builder);
       this.fishType = fishTypeIn;
+      this.fishTypeSupplier = () -> fishTypeIn;
    }
 
+   public FishBucketItem(java.util.function.Supplier<? extends EntityType<?>> fishTypeIn, java.util.function.Supplier<? extends Fluid> p_i49022_2_, Item.Settings builder) {
+      super(p_i49022_2_, builder);
+      this.fishType = null;
+      this.fishTypeSupplier = fishTypeIn;
+   }
+
    public void onEmptied(World worldIn, ItemStack p_203792_2_, BlockPos pos) {
       if (!worldIn.isClient) {
          this.spawnFish(worldIn, p_203792_2_, pos);
@@ -80,4 +88,9 @@
       }
 
    }
+
+   private final java.util.function.Supplier<? extends EntityType<?>> fishTypeSupplier;
+   protected EntityType<?> getFishType() {
+       return fishTypeSupplier.get();
+   }
 }
