--- a/net/minecraft/item/FlintAndSteelItem.java
+++ b/net/minecraft/item/FlintAndSteelItem.java
@@ -65,7 +65,8 @@
       boolean flag = false;
 
       for(Direction direction : Direction.Type.HORIZONTAL) {
-         if (worldIn.getBlockState(posIn.offset(direction)).getBlock() == Blocks.OBSIDIAN && ((NetherPortalBlock)Blocks.NETHER_PORTAL).createAreaHelper(worldIn, posIn) != null) {
+         BlockPos framePos = posIn.offset(direction);
+         if (worldIn.getBlockState(framePos).isPortalFrame(worldIn, framePos) && ((NetherPortalBlock)Blocks.NETHER_PORTAL).createAreaHelper(worldIn, posIn) != null) {
             flag = true;
          }
       }
