--- a/net/minecraft/item/HorseArmorItem.java
+++ b/net/minecraft/item/HorseArmorItem.java
@@ -8,17 +8,21 @@
 
 public class HorseArmorItem extends Item {
    private final int bonus;
-   private final String entityTexture;
+   private final Identifier texture;
 
    public HorseArmorItem(int p_i50042_1_, String p_i50042_2_, Item.Settings p_i50042_3_) {
+      this(p_i50042_1_, new Identifier("textures/entity/horse/armor/horse_armor_" + p_i50042_2_ + ".png"), p_i50042_3_);
+   }
+
+   public HorseArmorItem(int p_i50042_1_, Identifier texture, Item.Settings p_i50042_3_) {
       super(p_i50042_3_);
       this.bonus = p_i50042_1_;
-      this.entityTexture = "textures/entity/horse/armor/horse_armor_" + p_i50042_2_ + ".png";
+      this.texture = texture;
    }
 
    @Environment(EnvType.CLIENT)
    public Identifier getEntityTexture() {
-      return new Identifier(this.entityTexture);
+      return texture;
    }
 
    public int getBonus() {
