--- a/net/minecraft/item/LilyPadItem.java
+++ b/net/minecraft/item/LilyPadItem.java
@@ -50,7 +50,15 @@
             Material material = blockstate.getMaterial();
             FluidState ifluidstate = worldIn.getFluidState(blockpos);
             if ((ifluidstate.getFluid() == Fluids.WATER || material == Material.ICE) && worldIn.isAir(blockpos1)) {
+
+               // special case for handling block placement with water lilies
+               net.minecraftforge.common.util.BlockSnapshot blocksnapshot = net.minecraftforge.common.util.BlockSnapshot.getBlockSnapshot(worldIn, blockpos1);
                worldIn.setBlockState(blockpos1, Blocks.LILY_PAD.getDefaultState(), 11);
+               if (net.minecraftforge.event.ForgeEventFactory.onBlockPlace(playerIn, blocksnapshot, net.minecraft.util.math.Direction.UP)) {
+                  blocksnapshot.restore(true, false);
+                  return TypedActionResult.fail(itemstack);
+               }
+
                if (playerIn instanceof ServerPlayerEntity) {
                   Criterions.PLACED_BLOCK.trigger((ServerPlayerEntity)playerIn, blockpos1, itemstack);
                }
