--- a/net/minecraft/item/MinecartItem.java
+++ b/net/minecraft/item/MinecartItem.java
@@ -26,7 +26,7 @@
          double d2 = source.getZ() + (double)direction.getOffsetZ() * 1.125D;
          BlockPos blockpos = source.getBlockPos().offset(direction);
          BlockState blockstate = world.getBlockState(blockpos);
-         RailShape railshape = blockstate.getBlock() instanceof AbstractRailBlock ? blockstate.get(((AbstractRailBlock)blockstate.getBlock()).getShapeProperty()) : RailShape.NORTH_SOUTH;
+         RailShape railshape = blockstate.getBlock() instanceof AbstractRailBlock ? ((AbstractRailBlock)blockstate.getBlock()).getRailDirection(blockstate, world, blockpos, null) : RailShape.NORTH_SOUTH;
          double d3;
          if (blockstate.matches(BlockTags.RAILS)) {
             if (railshape.isAscending()) {
@@ -35,12 +35,12 @@
                d3 = 0.1D;
             }
          } else {
-            if (!blockstate.isAir() || !world.getBlockState(blockpos.down()).matches(BlockTags.RAILS)) {
+            if (!blockstate.isAir(world, blockpos) || !world.getBlockState(blockpos.down()).matches(BlockTags.RAILS)) {
                return this.defaultBehavior.dispense(source, stack);
             }
 
             BlockState blockstate1 = world.getBlockState(blockpos.down());
-            RailShape railshape1 = blockstate1.getBlock() instanceof AbstractRailBlock ? blockstate1.get(((AbstractRailBlock)blockstate1.getBlock()).getShapeProperty()) : RailShape.NORTH_SOUTH;
+            RailShape railshape1 = blockstate1.getBlock() instanceof AbstractRailBlock ? ((AbstractRailBlock)blockstate1.getBlock()).getRailDirection(blockstate1, world, blockpos.down(), null) : RailShape.NORTH_SOUTH;
             if (direction != Direction.DOWN && railshape1.isAscending()) {
                d3 = -0.4D;
             } else {
@@ -79,7 +79,7 @@
       } else {
          ItemStack itemstack = context.getStack();
          if (!world.isClient) {
-            RailShape railshape = blockstate.getBlock() instanceof AbstractRailBlock ? blockstate.get(((AbstractRailBlock)blockstate.getBlock()).getShapeProperty()) : RailShape.NORTH_SOUTH;
+            RailShape railshape = blockstate.getBlock() instanceof AbstractRailBlock ? ((AbstractRailBlock)blockstate.getBlock()).getRailDirection(blockstate, world, blockpos, null) : RailShape.NORTH_SOUTH;
             double d0 = 0.0D;
             if (railshape.isAscending()) {
                d0 = 0.5D;
