--- a/net/minecraft/world/poi/PointOfInterestType.java
+++ b/net/minecraft/world/poi/PointOfInterestType.java
@@ -19,7 +19,7 @@
 import net.minecraft.util.registry.Registry;
 import net.minecraft.village.VillagerProfession;
 
-public class PointOfInterestType {
+public class PointOfInterestType extends net.minecraftforge.registries.ForgeRegistryEntry<PointOfInterestType> {
    private static final Predicate<PointOfInterestType> IS_USED_BY_PROFESSION = (p_221041_0_) -> {
       return Registry.VILLAGER_PROFESSION.stream().map(VillagerProfession::getWorkStation).collect(Collectors.toSet()).contains(p_221041_0_);
    };
