--- a/net/minecraft/server/integrated/IntegratedServer.java
+++ b/net/minecraft/server/integrated/IntegratedServer.java
@@ -66,6 +66,8 @@
       this.upgradeWorld(saveName);
       WorldSaveHandler savehandler = this.getLevelStorage().createSaveHandler(saveName, this);
       this.loadWorldResourcePack(this.getLevelName(), savehandler);
+      // Move factory creation earlier to prevent startupquery deadlock
+      WorldGenerationProgressListener ichunkstatuslistener = this.worldGenerationProgressListenerFactory.create(11);
       LevelProperties worldinfo = savehandler.readProperties();
       if (worldinfo == null) {
          worldinfo = new LevelProperties(this.levelInfo, worldNameIn);
@@ -75,7 +77,6 @@
 
       worldinfo.method_24285(this.getServerModName(), this.method_24307().isPresent());
       this.loadWorldDataPacks(savehandler.getWorldDir(), worldinfo);
-      WorldGenerationProgressListener ichunkstatuslistener = this.worldGenerationProgressListenerFactory.create(11);
       this.createWorlds(savehandler, worldinfo, this.levelInfo, ichunkstatuslistener);
       if (this.getWorld(DimensionType.OVERWORLD).getLevelProperties().getDifficulty() == null) {
          this.setDifficulty(this.client.options.difficulty, true);
@@ -93,9 +94,10 @@
       this.setFlightEnabled(true);
       LOGGER.info("Generating keypair");
       this.setKeyPair(NetworkEncryptionUtils.generateServerKeyPair());
+      if (!net.minecraftforge.fml.server.ServerLifecycleHooks.handleServerAboutToStart(this)) return false;
       this.loadWorld(this.getLevelName(), this.getServerName(), this.levelInfo.getSeed(), this.levelInfo.getGeneratorType(), this.levelInfo.getGeneratorOptions());
       this.setMotd(this.getUserName() + " - " + this.getWorld(DimensionType.OVERWORLD).getLevelProperties().getLevelName());
-      return true;
+      return net.minecraftforge.fml.server.ServerLifecycleHooks.handleServerStarting(this);
    }
 
    public void tick(BooleanSupplier hasTimeLeft) {
@@ -130,6 +132,7 @@
    }
 
    public Difficulty getDefaultDifficulty() {
+      if (this.client.world == null) return this.client.options.difficulty; // Fix NPE just in case.
       return this.client.world.getLevelProperties().getDifficulty();
    }
 
@@ -221,6 +224,7 @@
    }
 
    public void stop(boolean waitForServer) {
+      if (isRunning())
       this.submitAndJoin(() -> {
          for(ServerPlayerEntity serverplayerentity : Lists.newArrayList(this.getPlayerManager().getPlayerList())) {
             if (!serverplayerentity.getUuid().equals(this.localPlayerUuid)) {
