--- a/net/minecraft/world/IWorld.java
+++ b/net/minecraft/world/IWorld.java
@@ -26,16 +26,16 @@
    long getSeed();
 
    default float getMoonSize() {
-      return Dimension.MOON_PHASE_TO_SIZE[this.getDimension().getMoonPhase(this.getLevelProperties().getTimeOfDay())];
+      return this.getDimension().getCurrentMoonPhaseFactor(this.getWorld().getTimeOfDay());
    }
 
    default float getSkyAngle(float partialTicks) {
-      return this.getDimension().getSkyAngle(this.getLevelProperties().getTimeOfDay(), partialTicks);
+      return this.getDimension().getSkyAngle(this.getWorld().getTimeOfDay(), partialTicks);
    }
 
    @Environment(EnvType.CLIENT)
    default int getMoonPhase() {
-      return this.getDimension().getMoonPhase(this.getLevelProperties().getTimeOfDay());
+      return this.getDimension().getMoonPhase(this.getWorld().getTimeOfDay());
    }
 
    TickScheduler<Block> getBlockTickScheduler();
