--- a/net/minecraft/block/BambooBlock.java
+++ b/net/minecraft/block/BambooBlock.java
@@ -90,10 +90,11 @@
       if (!state.canPlaceAt(worldIn, pos)) {
          worldIn.breakBlock(pos, true);
       } else if (state.get(STAGE) == 0) {
-         if (rand.nextInt(3) == 0 && worldIn.isAir(pos.up()) && worldIn.getBaseLightLevel(pos.up(), 0) >= 9) {
+         if (true) {
             int i = this.countBambooBelow(worldIn, pos) + 1;
-            if (i < 16) {
+            if (i < 16 && net.minecraftforge.common.ForgeHooks.onCropsGrowPre(worldIn, pos, state, rand.nextInt(3) == 0 && worldIn.isAir(pos.up()) && worldIn.getBaseLightLevel(pos.up(), 0) >= 9)) {
                this.updateLeaves(state, worldIn, pos, rand, i);
+               net.minecraftforge.common.ForgeHooks.onCropsGrowPost(worldIn, pos, state);
             }
          }
 
