--- a/net/minecraft/block/BedBlock.java
+++ b/net/minecraft/block/BedBlock.java
@@ -87,7 +87,9 @@
             }
          }
 
-         if (worldIn.dimension.canPlayersSleep() && worldIn.getBiome(pos) != Biomes.NETHER) {
+         net.minecraftforge.common.extensions.IForgeDimension.SleepResult sleepResult = worldIn.dimension.canSleepAt(player, pos);
+         if (sleepResult != net.minecraftforge.common.extensions.IForgeDimension.SleepResult.BED_EXPLODES) {
+            if (sleepResult == net.minecraftforge.common.extensions.IForgeDimension.SleepResult.DENY) return ActionResult.SUCCESS;
             if (state.get(OCCUPIED)) {
                if (!this.method_22357(worldIn, pos)) {
                   player.addChatMessage(new TranslatableText("block.minecraft.bed.occupied"), true);
