--- a/net/minecraft/block/BlockState.java
+++ b/net/minecraft/block/BlockState.java
@@ -51,7 +51,7 @@
 import net.minecraftforge.api.distmarker.Dist;
 import net.minecraftforge.api.distmarker.OnlyIn;
 
-public class BlockState extends AbstractState<Block, BlockState> implements State<BlockState> {
+public class BlockState extends AbstractState<Block, BlockState> implements State<BlockState>, net.minecraftforge.common.extensions.IForgeBlockState {
    @Nullable
    private BlockState.ShapeCache shapeCache;
    private final int luminance;
@@ -106,10 +106,14 @@
       return this.luminance;
    }
 
+   /** @deprecated use {@link BlockState#isAir(IBlockReader, BlockPos) */
+   @Deprecated
    public boolean isAir() {
       return this.getBlock().isAir(this);
    }
 
+   /** @deprecated use {@link BlockState#rotate(IWorld, BlockPos, Rotation) */
+   @Deprecated
    public MaterialColor getTopMaterialColor(BlockView worldIn, BlockPos pos) {
       return this.getBlock().getMapColor(this, worldIn, pos);
    }
