--- a/net/minecraft/block/CampfireBlock.java
+++ b/net/minecraft/block/CampfireBlock.java
@@ -188,7 +188,7 @@
          boolean flag = projectile instanceof AbstractFireballEntity || projectile instanceof ProjectileEntity && projectile.isOnFire();
          if (flag) {
             Entity entity = this.method_23756(projectile);
-            boolean flag1 = entity == null || entity instanceof PlayerEntity || worldIn.getGameRules().getBoolean(GameRules.MOB_GRIEFING);
+            boolean flag1 = entity == null || entity instanceof PlayerEntity || net.minecraftforge.event.ForgeEventFactory.getMobGriefingEvent(worldIn, entity);
             if (flag1 && !state.get(LIT) && !state.get(WATERLOGGED)) {
                BlockPos blockpos = hit.getBlockPos();
                worldIn.setBlockState(blockpos, state.with(Properties.LIT, Boolean.valueOf(true)), 11);
