--- a/net/minecraft/block/ComparatorBlock.java
+++ b/net/minecraft/block/ComparatorBlock.java
@@ -71,7 +71,7 @@
          blockstate = worldIn.getBlockState(blockpos);
          if (blockstate.hasComparatorOutput()) {
             i = blockstate.getComparatorOutput(worldIn, blockpos);
-         } else if (blockstate.isAir()) {
+         } else if (blockstate.isAir(worldIn, blockpos)) {
             ItemFrameEntity itemframeentity = this.getAttachedItemFrame(worldIn, direction, blockpos);
             if (itemframeentity != null) {
                i = itemframeentity.getComparatorPower();
@@ -157,4 +157,16 @@
    protected void appendProperties(StateManager.Builder<Block, BlockState> builder) {
       builder.add(FACING, MODE, POWERED);
    }
+
+   @Override
+   public boolean getWeakChanges(BlockState state, net.minecraft.world.WorldView world, BlockPos pos) {
+      return true;
+   }
+
+   @Override
+   public void onNeighborChange(BlockState state, net.minecraft.world.WorldView world, BlockPos pos, BlockPos neighbor) {
+      if (pos.getY() == neighbor.getY() && world instanceof World && !((World)world).isClient()) {
+         state.neighborUpdate((World)world, pos, world.getBlockState(neighbor).getBlock(), neighbor, false);
+      }
+   }
 }
