--- a/net/minecraft/block/DetectorRailBlock.java
+++ b/net/minecraft/block/DetectorRailBlock.java
@@ -129,6 +129,8 @@
 
    public int getComparatorOutput(BlockState blockState, World worldIn, BlockPos pos) {
       if (blockState.get(POWERED)) {
+         List<AbstractMinecartEntity> carts = this.getCarts(worldIn, pos, AbstractMinecartEntity.class, null);
+         if (!carts.isEmpty() && carts.get(0).getComparatorLevel() > -1) return carts.get(0).getComparatorLevel();
          List<CommandBlockMinecartEntity> list = this.getCarts(worldIn, pos, CommandBlockMinecartEntity.class, (Predicate<Entity>)null);
          if (!list.isEmpty()) {
             return list.get(0).getCommandExecutor().getSuccessCount();
