--- a/net/minecraft/block/FernBlock.java
+++ b/net/minecraft/block/FernBlock.java
@@ -8,7 +8,7 @@
 import net.minecraft.world.BlockView;
 import net.minecraft.world.World;
 
-public class FernBlock extends PlantBlock implements Fertilizable {
+public class FernBlock extends PlantBlock implements Fertilizable, net.minecraftforge.common.IShearable {
    protected static final VoxelShape SHAPE = Block.createCuboidShape(2.0D, 0.0D, 2.0D, 14.0D, 13.0D, 14.0D);
 
    protected FernBlock(Block.Settings properties) {
