--- a/net/minecraft/block/JigsawBlock.java
+++ b/net/minecraft/block/JigsawBlock.java
@@ -30,6 +30,13 @@
       return state.with(FACING, rot.rotate(state.get(FACING)));
    }
 
+   // FORGE: Fix jigsaw blocks not mirroring properly
+   @Override
+   @Deprecated
+   public BlockState mirror(BlockState state, net.minecraft.util.BlockMirror mirrorIn) {
+      return state.with(FACING, mirrorIn.apply(state.get(FACING)));
+   }
+
    public BlockState getPlacementState(ItemPlacementContext context) {
       return this.getDefaultState().with(FACING, context.getSide());
    }
