--- a/net/minecraft/block/NetherPortalBlock.java
+++ b/net/minecraft/block/NetherPortalBlock.java
@@ -71,7 +71,7 @@
 
    public boolean createPortalAt(IWorld worldIn, BlockPos pos) {
       NetherPortalBlock.AreaHelper netherportalblock$size = this.createAreaHelper(worldIn, pos);
-      if (netherportalblock$size != null) {
+      if (netherportalblock$size != null && !net.minecraftforge.event.ForgeEventFactory.onTrySpawnPortal(worldIn, pos, netherportalblock$size)) {
          netherportalblock$size.createPortal();
          return true;
       } else {
@@ -243,13 +243,13 @@
          int i;
          for(i = 0; i < 22; ++i) {
             BlockPos blockpos = pos.offset(directionIn, i);
-            if (!this.validStateInsidePortal(this.world.getBlockState(blockpos)) || this.world.getBlockState(blockpos.down()).getBlock() != Blocks.OBSIDIAN) {
+            if (!this.validStateInsidePortal(this.world.getBlockState(blockpos)) || !this.world.getBlockState(blockpos.down()).isPortalFrame(this.world, blockpos.down())) {
                break;
             }
          }
 
-         Block block = this.world.getBlockState(pos.offset(directionIn, i)).getBlock();
-         return block == Blocks.OBSIDIAN ? i : 0;
+         BlockPos framePos = pos.offset(directionIn, i);
+         return this.world.getBlockState(framePos).isPortalFrame(this.world, framePos) ? i : 0;
       }
 
       public int getHeight() {
@@ -276,13 +276,13 @@
                }
 
                if (i == 0) {
-                  block = this.world.getBlockState(blockpos.offset(this.positiveDir)).getBlock();
-                  if (block != Blocks.OBSIDIAN) {
+                  BlockPos framePos = blockpos.offset(this.positiveDir);
+                  if (!this.world.getBlockState(framePos).isPortalFrame(this.world, framePos)) {
                      break label56;
                   }
                } else if (i == this.width - 1) {
-                  block = this.world.getBlockState(blockpos.offset(this.negativeDir)).getBlock();
-                  if (block != Blocks.OBSIDIAN) {
+                  BlockPos framePos = blockpos.offset(this.negativeDir);
+                  if (!this.world.getBlockState(framePos).isPortalFrame(this.world, framePos)) {
                      break label56;
                   }
                }
@@ -290,7 +290,8 @@
          }
 
          for(int j = 0; j < this.width; ++j) {
-            if (this.world.getBlockState(this.lowerCorner.offset(this.negativeDir, j).up(this.height)).getBlock() != Blocks.OBSIDIAN) {
+            BlockPos framePos = this.lowerCorner.offset(this.negativeDir, j).up(this.height);
+            if (!this.world.getBlockState(framePos).isPortalFrame(this.world, framePos)) {
                this.height = 0;
                break;
             }
