--- a/net/minecraft/block/RedstoneWireBlock.java
+++ b/net/minecraft/block/RedstoneWireBlock.java
@@ -125,7 +125,7 @@
       BlockState blockstate1 = worldIn.getBlockState(blockpos1);
       if (!blockstate1.isSimpleFullBlock(worldIn, blockpos1)) {
          boolean flag = blockstate.isSideSolidFullSquare(worldIn, blockpos, Direction.UP) || blockstate.getBlock() == Blocks.HOPPER;
-         if (flag && connectsTo(worldIn.getBlockState(blockpos.up()))) {
+         if (flag && canConnectTo(worldIn.getBlockState(blockpos.up()), worldIn, blockpos.up(), null)) {
             if (blockstate.isFullCube(worldIn, blockpos)) {
                return WireConnection.UP;
             }
@@ -134,7 +134,7 @@
          }
       }
 
-      return !connectsTo(blockstate, face) && (blockstate.isSimpleFullBlock(worldIn, blockpos) || !connectsTo(worldIn.getBlockState(blockpos.down()))) ? WireConnection.NONE : WireConnection.SIDE;
+      return !canConnectTo(blockstate, worldIn, blockpos, face) && (blockstate.isSimpleFullBlock(worldIn, blockpos) || !canConnectTo(worldIn.getBlockState(blockpos.down()), worldIn, blockpos.down(), null)) ? WireConnection.NONE : WireConnection.SIDE;
    }
 
    public boolean canPlaceAt(BlockState state, WorldView worldIn, BlockPos pos) {
@@ -317,26 +317,18 @@
       boolean flag = blockstate.isSimpleFullBlock(worldIn, blockpos);
       BlockPos blockpos1 = pos.up();
       boolean flag1 = worldIn.getBlockState(blockpos1).isSimpleFullBlock(worldIn, blockpos1);
-      if (!flag1 && flag && connectsTo(worldIn, blockpos.up())) {
+      if (!flag1 && flag && canConnectTo(worldIn.getBlockState(blockpos.up()), worldIn, blockpos.up(), null)) {
          return true;
-      } else if (connectsTo(blockstate, side)) {
+      } else if (canConnectTo(blockstate, worldIn, blockpos, side)) {
          return true;
       } else if (blockstate.getBlock() == Blocks.REPEATER && blockstate.get(AbstractRedstoneGateBlock.POWERED) && blockstate.get(AbstractRedstoneGateBlock.FACING) == side) {
          return true;
       } else {
-         return !flag && connectsTo(worldIn, blockpos.down());
+         return !flag && canConnectTo(worldIn.getBlockState(blockpos.down()), worldIn, blockpos.down(), null);
       }
    }
 
-   protected static boolean connectsTo(BlockView worldIn, BlockPos pos) {
-      return connectsTo(worldIn.getBlockState(pos));
-   }
-
-   protected static boolean connectsTo(BlockState state) {
-      return connectsTo(state, (Direction)null);
-   }
-
-   protected static boolean connectsTo(BlockState blockState, @Nullable Direction side) {
+   protected static boolean canConnectTo(BlockState blockState, BlockView world, BlockPos pos, @Nullable Direction side) {
       Block block = blockState.getBlock();
       if (block == Blocks.REDSTONE_WIRE) {
          return true;
@@ -346,7 +338,7 @@
       } else if (Blocks.OBSERVER == blockState.getBlock()) {
          return side == blockState.get(ObserverBlock.FACING);
       } else {
-         return blockState.emitsRedstonePower() && side != null;
+         return blockState.canConnectRedstone(world, pos, side) && side != null;
       }
    }
 
