--- a/net/minecraft/block/SaplingBlock.java
+++ b/net/minecraft/block/SaplingBlock.java
@@ -29,6 +29,7 @@
 
    public void scheduledTick(BlockState state, ServerWorld worldIn, BlockPos pos, Random rand) {
       super.scheduledTick(state, worldIn, pos, rand);
+      if (!worldIn.isAreaLoaded(pos, 1)) return; // Forge: prevent loading unloaded chunks when checking neighbor's light
       if (worldIn.getLightLevel(pos.up()) >= 9 && rand.nextInt(7) == 0) {
          this.generate(worldIn, pos, state, rand);
       }
@@ -39,6 +40,7 @@
       if (p_226942_3_.get(STAGE) == 0) {
          p_226942_1_.setBlockState(p_226942_2_, p_226942_3_.cycle(STAGE), 4);
       } else {
+         if (!net.minecraftforge.event.ForgeEventFactory.saplingGrowTree(p_226942_1_, p_226942_4_, p_226942_2_)) return;
          this.generator.generate(p_226942_1_, p_226942_1_.getChunkManager().getChunkGenerator(), p_226942_2_, p_226942_3_, p_226942_4_);
       }
 
