--- a/net/minecraft/block/StemBlock.java
+++ b/net/minecraft/block/StemBlock.java
@@ -42,22 +42,24 @@
 
    public void scheduledTick(BlockState state, ServerWorld worldIn, BlockPos pos, Random rand) {
       super.scheduledTick(state, worldIn, pos, rand);
+      if (!worldIn.isAreaLoaded(pos, 1)) return; // Forge: prevent loading unloaded chunks when checking neighbor's light
       if (worldIn.getBaseLightLevel(pos, 0) >= 9) {
          float f = CropBlock.getAvailableMoisture(this, worldIn, pos);
-         if (rand.nextInt((int)(25.0F / f) + 1) == 0) {
+         if (net.minecraftforge.common.ForgeHooks.onCropsGrowPre(worldIn, pos, state, rand.nextInt((int)(25.0F / f) + 1) == 0)) {
             int i = state.get(AGE);
             if (i < 7) {
-               state = state.with(AGE, Integer.valueOf(i + 1));
-               worldIn.setBlockState(pos, state, 2);
+               worldIn.setBlockState(pos, state.with(AGE, Integer.valueOf(i + 1)), 2);
             } else {
                Direction direction = Direction.Type.HORIZONTAL.random(rand);
                BlockPos blockpos = pos.offset(direction);
-               Block block = worldIn.getBlockState(blockpos.down()).getBlock();
-               if (worldIn.getBlockState(blockpos).isAir() && (block == Blocks.FARMLAND || block == Blocks.DIRT || block == Blocks.COARSE_DIRT || block == Blocks.PODZOL || block == Blocks.GRASS_BLOCK)) {
+               BlockState soil = worldIn.getBlockState(blockpos.down());
+               Block block = soil.getBlock();
+               if (worldIn.isAir(blockpos) && (soil.canSustainPlant(worldIn, blockpos.down(), Direction.UP, this) || block == Blocks.FARMLAND || block == Blocks.DIRT || block == Blocks.COARSE_DIRT || block == Blocks.PODZOL || block == Blocks.GRASS_BLOCK)) {
                   worldIn.setBlockState(blockpos, this.gourdBlock.getDefaultState());
                   worldIn.setBlockState(pos, this.gourdBlock.getAttachedStem().getDefaultState().with(HorizontalFacingBlock.FACING, direction));
                }
             }
+            net.minecraftforge.common.ForgeHooks.onCropsGrowPost(worldIn, pos, state);
          }
 
       }
@@ -103,4 +105,10 @@
    public GourdBlock getGourdBlock() {
       return this.gourdBlock;
    }
+
+   //FORGE START
+   @Override
+   public net.minecraftforge.common.PlantType getPlantType(BlockView world, BlockPos pos) {
+      return net.minecraftforge.common.PlantType.Crop;
+   }
 }
