--- a/net/minecraft/block/SweetBerryBushBlock.java
+++ b/net/minecraft/block/SweetBerryBushBlock.java
@@ -49,8 +49,9 @@
    public void scheduledTick(BlockState state, ServerWorld worldIn, BlockPos pos, Random rand) {
       super.scheduledTick(state, worldIn, pos, rand);
       int i = state.get(AGE);
-      if (i < 3 && rand.nextInt(5) == 0 && worldIn.getBaseLightLevel(pos.up(), 0) >= 9) {
+      if (i < 3 && worldIn.getBaseLightLevel(pos.up(), 0) >= 9 && net.minecraftforge.common.ForgeHooks.onCropsGrowPre(worldIn, pos, state, rand.nextInt(5) == 0)) {
          worldIn.setBlockState(pos, state.with(AGE, Integer.valueOf(i + 1)), 2);
+         net.minecraftforge.common.ForgeHooks.onCropsGrowPost(worldIn, pos, state);
       }
 
    }
