--- a/net/minecraft/block/TallPlantBlock.java
+++ b/net/minecraft/block/TallPlantBlock.java
@@ -41,7 +41,7 @@
    @Nullable
    public BlockState getPlacementState(ItemPlacementContext context) {
       BlockPos blockpos = context.getBlockPos();
-      return blockpos.getY() < 255 && context.getWorld().getBlockState(blockpos.up()).canReplace(context) ? super.getPlacementState(context) : null;
+      return blockpos.getY() < context.getWorld().getDimension().getHeight() - 1 && context.getWorld().getBlockState(blockpos.up()).canReplace(context) ? super.getPlacementState(context) : null;
    }
 
    public void onPlaced(World worldIn, BlockPos pos, BlockState state, LivingEntity placer, ItemStack stack) {
@@ -53,6 +53,7 @@
          return super.canPlaceAt(state, worldIn, pos);
       } else {
          BlockState blockstate = worldIn.getBlockState(pos.down());
+         if (state.getBlock() != this) return super.canPlaceAt(state, worldIn, pos); //Forge: This function is called during world gen and placement, before this block is set, so if we are not 'here' then assume it's the pre-check.
          return blockstate.getBlock() == this && blockstate.get(HALF) == DoubleBlockHalf.LOWER;
       }
    }
