--- a/net/minecraft/block/TurtleEggBlock.java
+++ b/net/minecraft/block/TurtleEggBlock.java
@@ -141,7 +141,7 @@
       if (trampler instanceof TurtleEntity) {
          return false;
       } else {
-         return trampler instanceof LivingEntity && !(trampler instanceof PlayerEntity) ? worldIn.getGameRules().getBoolean(GameRules.MOB_GRIEFING) : true;
+         return trampler instanceof LivingEntity && !(trampler instanceof PlayerEntity) ? net.minecraftforge.event.ForgeEventFactory.getMobGriefingEvent(worldIn, trampler) : true;
       }
    }
 }
