--- a/net/minecraft/block/VineBlock.java
+++ b/net/minecraft/block/VineBlock.java
@@ -19,7 +19,7 @@
 import net.minecraft.world.IWorld;
 import net.minecraft.world.WorldView;
 
-public class VineBlock extends Block {
+public class VineBlock extends Block implements net.minecraftforge.common.IShearable {
    public static final BooleanProperty UP = ConnectingBlock.UP;
    public static final BooleanProperty NORTH = ConnectingBlock.NORTH;
    public static final BooleanProperty EAST = ConnectingBlock.EAST;
@@ -152,14 +152,14 @@
             worldIn.removeBlock(pos, false);
          }
 
-      } else if (worldIn.random.nextInt(4) == 0) {
+      } else if (worldIn.random.nextInt(4) == 0 && worldIn.isAreaLoaded(pos, 4)) { // Forge: check area to prevent loading unloaded chunks
          Direction direction = Direction.random(rand);
          BlockPos blockpos = pos.up();
          if (direction.getAxis().isHorizontal() && !state.get(getFacingProperty(direction))) {
             if (this.canGrowAt(worldIn, pos)) {
                BlockPos blockpos4 = pos.offset(direction);
                BlockState blockstate5 = worldIn.getBlockState(blockpos4);
-               if (blockstate5.isAir()) {
+               if (blockstate5.isAir(worldIn, blockpos4)) {
                   Direction direction3 = direction.rotateYClockwise();
                   Direction direction4 = direction.rotateYCounterclockwise();
                   boolean flag = state.get(getFacingProperty(direction3));
@@ -216,8 +216,8 @@
             if (pos.getY() > 0) {
                BlockPos blockpos1 = pos.down();
                BlockState blockstate1 = worldIn.getBlockState(blockpos1);
-               if (blockstate1.isAir() || blockstate1.getBlock() == this) {
-                  BlockState blockstate2 = blockstate1.isAir() ? this.getDefaultState() : blockstate1;
+               if (blockstate1.isAir(worldIn, blockpos1) || blockstate1.getBlock() == this) {
+                  BlockState blockstate2 = blockstate1.isAir(worldIn, blockpos1) ? this.getDefaultState() : blockstate1;
                   BlockState blockstate3 = this.getGrownState(state, blockstate2, rand);
                   if (blockstate2 != blockstate3 && this.hasHorizontalSide(blockstate3)) {
                      worldIn.setBlockState(blockpos1, blockstate3, 2);
@@ -322,4 +322,6 @@
    public static BooleanProperty getFacingProperty(Direction side) {
       return FACING_PROPERTIES.get(side);
    }
+
+   @Override public boolean isLadder(BlockState state, WorldView world, BlockPos pos, net.minecraft.entity.LivingEntity entity) { return true; }
 }
