--- a/net/minecraft/world/gen/surfacebuilder/SurfaceBuilder.java
+++ b/net/minecraft/world/gen/surfacebuilder/SurfaceBuilder.java
@@ -9,7 +9,7 @@
 import net.minecraft.world.biome.Biome;
 import net.minecraft.world.chunk.Chunk;
 
-public abstract class SurfaceBuilder<C extends SurfaceConfig> {
+public abstract class SurfaceBuilder<C extends SurfaceConfig> extends net.minecraftforge.registries.ForgeRegistryEntry<SurfaceBuilder<?>> {
    public static final BlockState AIR = Blocks.AIR.getDefaultState();
    public static final BlockState DIRT = Blocks.DIRT.getDefaultState();
    public static final BlockState GRASS_BLOCK = Blocks.GRASS_BLOCK.getDefaultState();
