--- a/net/minecraft/world/dimension/DimensionType.java
+++ b/net/minecraft/world/dimension/DimensionType.java
@@ -13,7 +13,7 @@
 import net.minecraft.world.biome.source.HorizontalVoronoiBiomeAccessType;
 import net.minecraft.world.biome.source.VoronoiBiomeAccessType;
 
-public class DimensionType implements DynamicSerializable {
+public class DimensionType extends net.minecraftforge.registries.ForgeRegistryEntry<DimensionType> implements DynamicSerializable {
    public static final DimensionType OVERWORLD = register("overworld", new DimensionType(1, "", "", OverworldDimension::new, true, HorizontalVoronoiBiomeAccessType.INSTANCE));
    public static final DimensionType THE_NETHER = register("the_nether", new DimensionType(0, "_nether", "DIM-1", TheNetherDimension::new, false, VoronoiBiomeAccessType.INSTANCE));
    public static final DimensionType THE_END = register("the_end", new DimensionType(2, "_end", "DIM1", TheEndDimension::new, false, VoronoiBiomeAccessType.INSTANCE));
@@ -23,18 +23,32 @@
    private final BiFunction<World, DimensionType, ? extends Dimension> factory;
    private final boolean hasSkyLight;
    private final BiomeAccessType biomeAccessType;
+   private final boolean isVanilla;
+   private final net.minecraftforge.common.ModDimension modType;
+   private final net.minecraft.util.PacketByteBuf data;
 
    private static DimensionType register(String key, DimensionType type) {
       return Registry.register(Registry.DIMENSION_TYPE, type.id, key, type);
    }
 
+   //Forge, Internal use only. Use DimensionManager instead.
+   @Deprecated
    protected DimensionType(int p_i225789_1_, String p_i225789_2_, String p_i225789_3_, BiFunction<World, DimensionType, ? extends Dimension> p_i225789_4_, boolean p_i225789_5_, BiomeAccessType p_i225789_6_) {
+      this(p_i225789_1_, p_i225789_2_, p_i225789_3_, p_i225789_4_, p_i225789_5_, p_i225789_6_, null, null);
+   }
+
+   //Forge, Internal use only. Use DimensionManager instead.
+   @Deprecated
+   public DimensionType(int p_i225789_1_, String p_i225789_2_, String p_i225789_3_, BiFunction<World, DimensionType, ? extends Dimension> p_i225789_4_, boolean p_i225789_5_, BiomeAccessType p_i225789_6_, @Nullable net.minecraftforge.common.ModDimension modType, @Nullable net.minecraft.util.PacketByteBuf data) {
       this.id = p_i225789_1_;
       this.suffix = p_i225789_2_;
       this.saveDir = p_i225789_3_;
       this.factory = p_i225789_4_;
       this.hasSkyLight = p_i225789_5_;
       this.biomeAccessType = p_i225789_6_;
+      this.isVanilla = this.id >= 0 && this.id <= 2;
+      this.modType = modType;
+      this.data = data;
    }
 
    public static DimensionType deserialize(Dynamic<?> p_218271_0_) {
@@ -49,8 +63,9 @@
       return this.id + -1;
    }
 
+   @Deprecated //Forge Do not use, only used for villages backwards compatibility
    public String getSuffix() {
-      return this.suffix;
+      return isVanilla ? this.suffix : "";
    }
 
    public File getSaveDirectory(File p_212679_1_) {
@@ -62,7 +77,7 @@
    }
 
    public String toString() {
-      return getId(this).toString();
+      return "DimensionType{" + getId(this) + "}";
    }
 
    @Nullable
@@ -70,7 +85,21 @@
       return Registry.DIMENSION_TYPE.get(id - -1);
    }
 
+   public boolean isVanilla() {
+      return this.isVanilla;
+   }
+
    @Nullable
+   public net.minecraftforge.common.ModDimension getModType() {
+      return this.modType;
+   }
+
+   @Nullable
+   public net.minecraft.util.PacketByteBuf getData() {
+      return this.data;
+   }
+
+   @Nullable
    public static DimensionType byId(Identifier nameIn) {
       return Registry.DIMENSION_TYPE.get(nameIn);
    }
