--- a/net/minecraft/world/dimension/OverworldDimension.java
+++ b/net/minecraft/world/dimension/OverworldDimension.java
@@ -53,10 +53,6 @@
       super(worldIn, typeIn, 0.0F);
    }
 
-   public DimensionType getType() {
-      return DimensionType.OVERWORLD;
-   }
-
    public ChunkGenerator<? extends ChunkGeneratorConfig> createChunkGenerator() {
       LevelGeneratorType worldtype = this.world.getLevelProperties().getGeneratorType();
       ChunkGeneratorType<FlatChunkGeneratorConfig, FlatChunkGenerator> chunkgeneratortype = ChunkGeneratorType.FLAT;
