--- a/net/minecraft/world/gen/feature/DarkOakTreeFeature.java
+++ b/net/minecraft/world/gen/feature/DarkOakTreeFeature.java
@@ -20,17 +20,17 @@
       int j = positionIn.getX();
       int k = positionIn.getY();
       int l = positionIn.getZ();
-      if (k >= 1 && k + i + 1 < 256) {
+      if (k >= 1 && k + i + 1 < generationReader.getMaxHeight()) {
          BlockPos blockpos = positionIn.down();
-         if (!isNaturalDirtOrGrass(generationReader, blockpos)) {
+         if (!isSoil(generationReader, blockpos, configIn.getSapling())) {
             return false;
          } else if (!this.doesTreeFit(generationReader, positionIn, i)) {
             return false;
          } else {
-            this.setToDirt(generationReader, blockpos);
-            this.setToDirt(generationReader, blockpos.east());
-            this.setToDirt(generationReader, blockpos.south());
-            this.setToDirt(generationReader, blockpos.south().east());
+            this.setDirtAt(generationReader, blockpos, positionIn);
+            this.setDirtAt(generationReader, blockpos.east(), positionIn);
+            this.setDirtAt(generationReader, blockpos.south(), positionIn);
+            this.setDirtAt(generationReader, blockpos.south().east(), positionIn);
             Direction direction = Direction.Type.HORIZONTAL.random(rand);
             int i1 = i - rand.nextInt(4);
             int j1 = 2 - rand.nextInt(3);
