--- a/net/minecraft/world/gen/feature/OreFeatureConfig.java
+++ b/net/minecraft/world/gen/feature/OreFeatureConfig.java
@@ -34,7 +34,7 @@
       return new OreFeatureConfig(orefeatureconfig$fillerblocktype, blockstate, i);
    }
 
-   public static enum Target {
+   public static enum Target implements net.minecraftforge.common.IExtensibleEnum {
       NATURAL_STONE("natural_stone", (p_214739_0_) -> {
          if (p_214739_0_ == null) {
             return false;
@@ -67,5 +67,15 @@
       public Predicate<BlockState> getCondition() {
          return this.predicate;
       }
+      
+      public static Target create(String enumName, String nameIn, Predicate<BlockState> predicateIn) {
+         throw new IllegalStateException("Enum not extended");
+      }
+
+      @Override
+      @Deprecated
+      public void init() {
+          nameMap.put(getName(), this);
+      }
    }
 }
