--- a/net/minecraft/world/gen/feature/SpringFeature.java
+++ b/net/minecraft/world/gen/feature/SpringFeature.java
@@ -21,7 +21,7 @@
          return false;
       } else {
          BlockState blockstate = worldIn.getBlockState(pos);
-         if (!blockstate.isAir() && !config.validBlocks.contains(blockstate.getBlock())) {
+         if (!blockstate.isAir(worldIn, pos) && !config.validBlocks.contains(blockstate.getBlock())) {
             return false;
          } else {
             int i = 0;
