--- a/net/minecraft/world/gen/feature/VinesFeature.java
+++ b/net/minecraft/world/gen/feature/VinesFeature.java
@@ -21,7 +21,7 @@
    public boolean generate(IWorld worldIn, ChunkGenerator<? extends ChunkGeneratorConfig> generator, Random rand, BlockPos pos, DefaultFeatureConfig config) {
       BlockPos.Mutable blockpos$mutable = new BlockPos.Mutable(pos);
 
-      for(int i = pos.getY(); i < 256; ++i) {
+      for(int i = pos.getY(); i < worldIn.getWorld().getDimension().getHeight(); ++i) {
          blockpos$mutable.set(pos);
          blockpos$mutable.setOffset(rand.nextInt(4) - rand.nextInt(4), 0, rand.nextInt(4) - rand.nextInt(4));
          blockpos$mutable.setY(i);
