--- a/net/minecraft/entity/ai/pathing/LandPathNodeMaker.java
+++ b/net/minecraft/entity/ai/pathing/LandPathNodeMaker.java
@@ -384,16 +384,9 @@
             for(int j = -1; j <= 1; ++j) {
                for(int k = -1; k <= 1; ++k) {
                   if (i != 0 || k != 0) {
-                     Block block = p_193578_0_.getBlockState(blockpos$pooledmutable.set(i + blockaccessIn, j + x, k + y)).getBlock();
-                     if (block == Blocks.CACTUS) {
-                        z = PathNodeType.DANGER_CACTUS;
-                     } else if (block != Blocks.FIRE && block != Blocks.LAVA) {
-                        if (block == Blocks.SWEET_BERRY_BUSH) {
-                           z = PathNodeType.DANGER_OTHER;
-                        }
-                     } else {
-                        z = PathNodeType.DANGER_FIRE;
-                     }
+                     PathNodeType type = getBasicPathNodeType(p_193578_0_, blockaccessIn + i, x + j, y + k).getDanger();
+                     if (type != null)
+                        z = type;
                   }
                }
             }
@@ -406,9 +399,11 @@
    protected static PathNodeType getBasicPathNodeType(BlockView p_189553_0_, int blockaccessIn, int x, int y) {
       BlockPos blockpos = new BlockPos(blockaccessIn, x, y);
       BlockState blockstate = p_189553_0_.getBlockState(blockpos);
+      PathNodeType type = blockstate.getAiPathNodeType(p_189553_0_, blockpos);
+      if (type != null) return type;
       Block block = blockstate.getBlock();
       Material material = blockstate.getMaterial();
-      if (blockstate.isAir()) {
+      if (blockstate.isAir(p_189553_0_, blockpos)) {
          return PathNodeType.OPEN;
       } else if (!block.matches(BlockTags.TRAPDOORS) && block != Blocks.LILY_PAD) {
          if (block == Blocks.FIRE) {
