--- a/net/minecraft/block/entity/AbstractFurnaceBlockEntity.java
+++ b/net/minecraft/block/entity/AbstractFurnaceBlockEntity.java
@@ -86,6 +86,7 @@
       this.recipeType = recipeTypeIn;
    }
 
+   @Deprecated //Forge - get burn times by calling ForgeHooks#getBurnTime(ItemStack)
    public static Map<Item, Integer> createFuelTimeMap() {
       Map<Item, Integer> map = Maps.newLinkedHashMap();
       addFuel(map, Items.LAVA_BUCKET, 20000);
@@ -99,18 +100,8 @@
       addFuel(map, ItemTags.WOODEN_SLABS, 150);
       addFuel(map, ItemTags.WOODEN_TRAPDOORS, 300);
       addFuel(map, ItemTags.WOODEN_PRESSURE_PLATES, 300);
-      addFuel(map, Blocks.OAK_FENCE, 300);
-      addFuel(map, Blocks.BIRCH_FENCE, 300);
-      addFuel(map, Blocks.SPRUCE_FENCE, 300);
-      addFuel(map, Blocks.JUNGLE_FENCE, 300);
-      addFuel(map, Blocks.DARK_OAK_FENCE, 300);
-      addFuel(map, Blocks.ACACIA_FENCE, 300);
-      addFuel(map, Blocks.OAK_FENCE_GATE, 300);
-      addFuel(map, Blocks.BIRCH_FENCE_GATE, 300);
-      addFuel(map, Blocks.SPRUCE_FENCE_GATE, 300);
-      addFuel(map, Blocks.JUNGLE_FENCE_GATE, 300);
-      addFuel(map, Blocks.DARK_OAK_FENCE_GATE, 300);
-      addFuel(map, Blocks.ACACIA_FENCE_GATE, 300);
+      addFuel(map, net.minecraftforge.common.Tags.Items.FENCES_WOODEN, 300);
+      addFuel(map, net.minecraftforge.common.Tags.Items.FENCE_GATES_WOODEN, 300);
       addFuel(map, Blocks.NOTE_BLOCK, 300);
       addFuel(map, Blocks.BOOKSHELF, 300);
       addFuel(map, Blocks.LECTERN, 300);
@@ -170,9 +161,9 @@
       super.fromTag(compound);
       this.inventory = DefaultedList.ofSize(this.getInvSize(), ItemStack.EMPTY);
       Inventories.fromTag(compound, this.inventory);
-      this.burnTime = compound.getShort("BurnTime");
-      this.cookTime = compound.getShort("CookTime");
-      this.cookTimeTotal = compound.getShort("CookTimeTotal");
+      this.burnTime = compound.getInt("BurnTime");
+      this.cookTime = compound.getInt("CookTime");
+      this.cookTimeTotal = compound.getInt("CookTimeTotal");
       this.fuelTime = this.getFuelTime(this.inventory.get(1));
       int i = compound.getShort("RecipesUsedSize");
 
@@ -186,9 +177,9 @@
 
    public CompoundTag toTag(CompoundTag compound) {
       super.toTag(compound);
-      compound.putShort("BurnTime", (short)this.burnTime);
-      compound.putShort("CookTime", (short)this.cookTime);
-      compound.putShort("CookTimeTotal", (short)this.cookTimeTotal);
+      compound.putInt("BurnTime", this.burnTime);
+      compound.putInt("CookTime", this.cookTime);
+      compound.putInt("CookTimeTotal", this.cookTimeTotal);
       Inventories.toTag(compound, this.inventory);
       compound.putShort("RecipesUsedSize", (short)this.recipesUsed.size());
       int i = 0;
@@ -218,12 +209,14 @@
                this.fuelTime = this.burnTime;
                if (this.isBurning()) {
                   flag1 = true;
+                  if (itemstack.hasContainerItem())
+                      this.inventory.set(1, itemstack.getContainerItem());
+                  else
                   if (!itemstack.isEmpty()) {
                      Item item = itemstack.getItem();
                      itemstack.decrement(1);
                      if (itemstack.isEmpty()) {
-                        Item item1 = item.getRecipeRemainder();
-                        this.inventory.set(1, item1 == null ? ItemStack.EMPTY : new ItemStack(item1));
+                        this.inventory.set(1, itemstack.getContainerItem());
                      }
                   }
                }
@@ -267,10 +260,10 @@
                return true;
             } else if (!itemstack1.isItemEqualIgnoreDamage(itemstack)) {
                return false;
-            } else if (itemstack1.getCount() < this.getInvMaxStackAmount() && itemstack1.getCount() < itemstack1.getMaxCount()) {
+            } else if (itemstack1.getCount() + itemstack.getCount() <= this.getInvMaxStackAmount() && itemstack1.getCount() + itemstack.getCount() <= itemstack1.getMaxCount()) { // Forge fix: make furnace respect stack sizes in furnace recipes
                return true;
             } else {
-               return itemstack1.getCount() < itemstack.getMaxCount();
+               return itemstack1.getCount() + itemstack.getCount() <= itemstack.getMaxCount(); // Forge fix: make furnace respect stack sizes in furnace recipes
             }
          }
       } else {
@@ -286,7 +279,7 @@
          if (itemstack2.isEmpty()) {
             this.inventory.set(2, itemstack1.copy());
          } else if (itemstack2.getItem() == itemstack1.getItem()) {
-            itemstack2.increment(1);
+            itemstack2.increment(itemstack1.getCount());
          }
 
          if (!this.world.isClient) {
@@ -306,7 +299,7 @@
          return 0;
       } else {
          Item item = fuel.getItem();
-         return createFuelTimeMap().getOrDefault(item, 0);
+         return net.minecraftforge.common.ForgeHooks.getBurnTime(fuel);
       }
    }
 
@@ -315,7 +308,7 @@
    }
 
    public static boolean canUseAsFuel(ItemStack stack) {
-      return createFuelTimeMap().containsKey(stack.getItem());
+      return net.minecraftforge.common.ForgeHooks.getBurnTime(stack) > 0;
    }
 
    public int[] getInvAvailableSlots(Direction side) {
@@ -463,4 +456,27 @@
       }
 
    }
+
+   net.minecraftforge.common.util.LazyOptional<? extends net.minecraftforge.items.IItemHandler>[] handlers =
+           net.minecraftforge.items.wrapper.SidedInvWrapper.create(this, Direction.UP, Direction.DOWN, Direction.NORTH);
+
+   @Override
+   public <T> net.minecraftforge.common.util.LazyOptional<T> getCapability(net.minecraftforge.common.capabilities.Capability<T> capability, @Nullable Direction facing) {
+      if (!this.removed && facing != null && capability == net.minecraftforge.items.CapabilityItemHandler.ITEM_HANDLER_CAPABILITY) {
+         if (facing == Direction.UP)
+            return handlers[0].cast();
+         else if (facing == Direction.DOWN)
+            return handlers[1].cast();
+         else
+            return handlers[2].cast();
+      }
+      return super.getCapability(capability, facing);
+   }
+
+   @Override
+   public void markRemoved() {
+      super.markRemoved();
+      for (int x = 0; x < handlers.length; x++)
+        handlers[x].invalidate();
+   }
 }
