--- a/net/minecraft/network/packet/s2c/play/CommandTreeS2CPacket.java
+++ b/net/minecraft/network/packet/s2c/play/CommandTreeS2CPacket.java
@@ -112,6 +112,9 @@
          String s = buf.readString(32767);
          ArgumentType<?> argumenttype = ArgumentTypes.fromPacket(buf);
          if (argumenttype == null) {
+            if ((flags & 16) != 0) { // FORGE: Flush unused suggestion data
+               buf.readIdentifier();
+            }
             return null;
          } else {
             RequiredArgumentBuilder<CommandSource, ?> requiredargumentbuilder = RequiredArgumentBuilder.argument(s, argumenttype);
