--- a/net/minecraft/network/packet/s2c/play/PlayerRespawnS2CPacket.java
+++ b/net/minecraft/network/packet/s2c/play/PlayerRespawnS2CPacket.java
@@ -17,6 +17,7 @@
    private long sha256Seed;
    private GameMode gameMode;
    private LevelGeneratorType generatorType;
+   private int dimensionInt;
 
    public PlayerRespawnS2CPacket() {
    }
@@ -33,7 +34,7 @@
    }
 
    public void read(PacketByteBuf buf) throws IOException {
-      this.dimension = DimensionType.byRawId(buf.readInt());
+      this.dimensionInt = buf.readInt();
       this.sha256Seed = buf.readLong();
       this.gameMode = GameMode.byId(buf.readUnsignedByte());
       this.generatorType = LevelGeneratorType.getTypeFromName(buf.readString(16));
@@ -52,7 +53,7 @@
 
    @Environment(EnvType.CLIENT)
    public DimensionType getDimension() {
-      return this.dimension;
+      return this.dimension == null ? this.dimension = net.minecraftforge.fml.network.NetworkHooks.getDummyDimType(this.dimensionInt) : this.dimension;
    }
 
    @Environment(EnvType.CLIENT)
