--- a/net/minecraft/network/packet/s2c/login/LoginSuccessS2CPacket.java
+++ b/net/minecraft/network/packet/s2c/login/LoginSuccessS2CPacket.java
@@ -24,7 +24,7 @@
    public void read(PacketByteBuf buf) throws IOException {
       String s = buf.readString(36);
       String s1 = buf.readString(16);
-      UUID uuid = UUID.fromString(s);
+      UUID uuid = s.length() > 0 ? UUID.fromString(s) : null; // Forge: prevent exception with bad data.
       this.profile = new GameProfile(uuid, s1);
    }
 
