--- a/net/minecraft/container/BeaconContainer.java
+++ b/net/minecraft/container/BeaconContainer.java
@@ -17,7 +17,7 @@
 public class BeaconContainer extends Container {
    private final Inventory paymentInv = new BasicInventory(1) {
       public boolean isValidInvStack(int index, ItemStack stack) {
-         return stack.getItem() == Items.EMERALD || stack.getItem() == Items.DIAMOND || stack.getItem() == Items.GOLD_INGOT || stack.getItem() == Items.IRON_INGOT;
+         return stack.isBeaconPayment();
       }
 
       public int getInvMaxStackAmount() {
@@ -87,10 +87,8 @@
             }
 
             slot.onStackChanged(itemstack1, itemstack);
-         } else if (!this.paymentSlot.hasStack() && this.paymentSlot.canInsert(itemstack1) && itemstack1.getCount() == 1) {
-            if (!this.insertItem(itemstack1, 0, 1, false)) {
-               return ItemStack.EMPTY;
-            }
+         } else if (this.insertItem(itemstack1, 0, 1, false)) { //Forge Fix Shift Clicking in beacons with stacks larger then 1.
+            return ItemStack.EMPTY;
          } else if (index >= 1 && index < 28) {
             if (!this.insertItem(itemstack1, 28, 37, false)) {
                return ItemStack.EMPTY;
@@ -156,8 +154,7 @@
       }
 
       public boolean canInsert(ItemStack stack) {
-         Item item = stack.getItem();
-         return item == Items.EMERALD || item == Items.DIAMOND || item == Items.GOLD_INGOT || item == Items.IRON_INGOT;
+         return stack.isBeaconPayment();
       }
 
       public int getMaxStackAmount() {
