--- a/net/minecraft/world/biome/layer/BiomeLayers.java
+++ b/net/minecraft/world/biome/layer/BiomeLayers.java
@@ -59,13 +59,11 @@
       iareafactory = AddDeepOceanLayer.INSTANCE.create(p_227475_2_.apply(4L), iareafactory);
       iareafactory = stack(1000L, ScaleLayer.NORMAL, iareafactory, 0, p_227475_2_);
       int i = p_227475_0_ == LevelGeneratorType.LARGE_BIOMES ? 6 : p_227475_1_.getBiomeSize();
+      i = getModdedBiomeSize(p_227475_0_, i);
       int j = p_227475_1_.getRiverSize();
       LayerFactory<T> lvt_7_1_ = stack(1000L, ScaleLayer.NORMAL, iareafactory, 0, p_227475_2_);
       lvt_7_1_ = SimpleLandNoiseLayer.INSTANCE.create((LayerSampleContext)p_227475_2_.apply(100L), lvt_7_1_);
-      LayerFactory<T> lvt_8_1_ = (new SetBaseBiomesLayer(p_227475_0_, p_227475_1_.getForcedBiome())).create(p_227475_2_.apply(200L), iareafactory);
-      lvt_8_1_ = AddBambooJungleLayer.INSTANCE.create((LayerSampleContext)p_227475_2_.apply(1001L), lvt_8_1_);
-      lvt_8_1_ = stack(1000L, ScaleLayer.NORMAL, lvt_8_1_, 2, p_227475_2_);
-      lvt_8_1_ = EaseBiomeEdgeLayer.INSTANCE.create((LayerSampleContext)p_227475_2_.apply(1000L), lvt_8_1_);
+      LayerFactory<T> lvt_8_1_ = p_227475_0_.getBiomeLayer(iareafactory, p_227475_1_, p_227475_2_);
       LayerFactory<T> lvt_9_1_ = stack(1000L, ScaleLayer.NORMAL, lvt_7_1_, 2, p_227475_2_);
       lvt_8_1_ = AddHillsLayer.INSTANCE.create((LayerSampleContext)p_227475_2_.apply(1000L), lvt_8_1_, lvt_9_1_);
       lvt_7_1_ = stack(1000L, ScaleLayer.NORMAL, lvt_7_1_, 2, p_227475_2_);
@@ -121,6 +119,15 @@
       }
    }
 
+   /* ======================================== FORGE START =====================================*/
+   public static int getModdedBiomeSize(LevelGeneratorType worldType, int original)
+   {
+       net.minecraftforge.event.terraingen.WorldTypeEvent.BiomeSize event = new net.minecraftforge.event.terraingen.WorldTypeEvent.BiomeSize(worldType, original);
+       net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(event);
+       return event.getNewSize();
+   }
+   /* ========================================= FORGE END ======================================*/
+
    protected static boolean isOcean(int biomeIn) {
       return biomeIn == WARM_OCEAN_ID || biomeIn == LUKEWARM_OCEAN_ID || biomeIn == OCEAN_ID || biomeIn == COLD_OCEAN_ID || biomeIn == FROZEN_OCEAN_ID || biomeIn == DEEP_WARM_OCEAN_ID || biomeIn == DEEP_LUKEWARM_OCEAN_ID || biomeIn == DEEP_OCEAN_ID || biomeIn == DEEP_COLD_OCEAN_ID || biomeIn == DEEP_FROZEN_OCEAN_ID;
    }
