--- a/net/minecraft/client/gui/hud/InGameHud.java
+++ b/net/minecraft/client/gui/hud/InGameHud.java
@@ -405,6 +405,9 @@
 
          for(StatusEffectInstance effectinstance : Ordering.natural().reverse().sortedCopy(collection)) {
             StatusEffect effect = effectinstance.getEffectType();
+            if (!effectinstance.shouldRenderHUD()) continue;
+            // Rebind in case previous renderHUDEffect changed texture
+            this.client.getTextureManager().bindTexture(ContainerScreen.BACKGROUND_TEXTURE);
             if (effectinstance.shouldShowIcon()) {
                int k = this.scaledWidth;
                int l = 1;
@@ -442,6 +445,7 @@
                   RenderSystem.color4f(1.0F, 1.0F, 1.0F, f1);
                   blit(j1 + 3, k1 + 3, this.getBlitOffset(), 18, 18, textureatlassprite);
                });
+               effectinstance.renderHUDEffect(this, k, l, this.getBlitOffset(), f);
             }
          }
 
@@ -567,6 +571,7 @@
          }
 
          String s = itextcomponent.asFormattedString();
+         s = this.currentStack.getHighlightTip(s);
          int i = (this.scaledWidth - this.getFontRenderer().getStringWidth(s)) / 2;
          int j = this.scaledHeight - 59;
          if (!this.client.interactionManager.hasStatusBars()) {
@@ -583,7 +588,13 @@
             RenderSystem.enableBlend();
             RenderSystem.defaultBlendFunc();
             fill(i - 2, j - 2, i + this.getFontRenderer().getStringWidth(s) + 2, j + 9 + 2, this.client.options.getTextBackgroundColor(0));
+            TextRenderer font = currentStack.getItem().getFontRenderer(currentStack);
+            if (font == null) {
             this.getFontRenderer().drawWithShadow(s, (float)i, (float)j, 16777215 + (k << 24));
+            } else {
+                i = (this.scaledWidth - font.getStringWidth(s)) / 2;
+                font.drawWithShadow(s, (float)i, (float)j, 16777215 + (k << 24));
+            }
             RenderSystem.disableBlend();
             RenderSystem.popMatrix();
          }
@@ -1044,7 +1055,7 @@
          ItemStack itemstack = this.client.player.inventory.getMainHandStack();
          if (itemstack.isEmpty()) {
             this.heldItemTooltipFade = 0;
-         } else if (!this.currentStack.isEmpty() && itemstack.getItem() == this.currentStack.getItem() && itemstack.getName().equals(this.currentStack.getName())) {
+         } else if (!this.currentStack.isEmpty() && itemstack.getItem() == this.currentStack.getItem() && (itemstack.getName().equals(this.currentStack.getName()) && itemstack.getHighlightTip(itemstack.getName().asString()).equals(currentStack.getHighlightTip(currentStack.getName().asString())))) {
             if (this.heldItemTooltipFade > 0) {
                --this.heldItemTooltipFade;
             }
