--- a/net/minecraft/world/gen/foliage/FoliagePlacerType.java
+++ b/net/minecraft/world/gen/foliage/FoliagePlacerType.java
@@ -4,7 +4,7 @@
 import java.util.function.Function;
 import net.minecraft.util.registry.Registry;
 
-public class FoliagePlacerType<P extends FoliagePlacer> {
+public class FoliagePlacerType<P extends FoliagePlacer> extends net.minecraftforge.registries.ForgeRegistryEntry<FoliagePlacerType<?>> {
    public static final FoliagePlacerType<BlobFoliagePlacer> BLOB_FOLIAGE_PLACER = register("blob_foliage_placer", BlobFoliagePlacer::new);
    public static final FoliagePlacerType<SpruceFoliagePlacer> SPRUCE_FOLIAGE_PLACER = register("spruce_foliage_placer", SpruceFoliagePlacer::new);
    public static final FoliagePlacerType<PineFoliagePlacer> PINE_FOLIAGE_PLACER = register("pine_foliage_placer", PineFoliagePlacer::new);
