--- a/net/minecraft/world/biome/Biomes.java
+++ b/net/minecraft/world/biome/Biomes.java
@@ -4,8 +4,10 @@
 import net.minecraft.util.Identifier;
 import net.minecraft.util.registry.Registry;
 
+@net.minecraftforge.registries.ObjectHolder("minecraft")
 public abstract class Biomes {
    public static final Biome OCEAN = register(0, "ocean", new OceanBiome());
+   @net.minecraftforge.registries.ObjectHolder("minecraft:ocean")
    public static final Biome DEFAULT = OCEAN;
    public static final Biome PLAINS = register(1, "plains", new PlainsBiome());
    public static final Biome DESERT = register(2, "desert", new DesertBiome());
