--- a/net/minecraft/entity/mob/EvokerEntity.java
+++ b/net/minecraft/entity/mob/EvokerEntity.java
@@ -281,7 +281,7 @@
             return false;
          } else if (EvokerEntity.this.age < this.startTime) {
             return false;
-         } else if (!EvokerEntity.this.world.getGameRules().getBoolean(GameRules.MOB_GRIEFING)) {
+         } else if (!net.minecraftforge.event.ForgeEventFactory.getMobGriefingEvent(EvokerEntity.this.world, EvokerEntity.this)) {
             return false;
          } else {
             List<SheepEntity> list = EvokerEntity.this.world.getTargets(SheepEntity.class, this.purpleSheepPredicate, EvokerEntity.this, EvokerEntity.this.getBoundingBox().expand(16.0D, 4.0D, 16.0D));
