--- a/net/minecraft/entity/Entity.java
+++ b/net/minecraft/entity/Entity.java
@@ -62,7 +62,6 @@
 import net.minecraft.sound.SoundCategory;
 import net.minecraft.sound.SoundEvent;
 import net.minecraft.sound.SoundEvents;
-import net.minecraft.tag.BlockTags;
 import net.minecraft.tag.FluidTags;
 import net.minecraft.tag.Tag;
 import net.minecraft.text.ClickEvent;
@@ -103,12 +102,13 @@
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
 
-public abstract class Entity implements Nameable, CommandOutput {
+public abstract class Entity extends net.minecraftforge.common.capabilities.CapabilityProvider<Entity> implements Nameable, CommandOutput, net.minecraftforge.common.extensions.IForgeEntity {
    protected static final Logger LOGGER = LogManager.getLogger();
    private static final AtomicInteger MAX_ENTITY_ID = new AtomicInteger();
    private static final List<ItemStack> EMPTY_STACK_LIST = Collections.emptyList();
    private static final Box NULL_BOX = new Box(0.0D, 0.0D, 0.0D, 0.0D, 0.0D, 0.0D);
    private static double renderDistanceMultiplier = 1.0D;
+   @Deprecated // Forge: Use the getter to allow overriding in mods
    private final EntityType<?> type;
    private int entityId = MAX_ENTITY_ID.incrementAndGet();
    public boolean inanimate;
@@ -136,6 +136,7 @@
    public boolean collided;
    public boolean velocityModified;
    protected Vec3d movementMultiplier = Vec3d.ZERO;
+   @Deprecated //Forge: Use isAlive, remove(boolean) and revive() instead of directly accessing this field. To allow the entity to react to and better control this information.
    public boolean removed;
    public float prevHorizontalSpeed;
    public float horizontalSpeed;
@@ -194,6 +195,7 @@
    private float standingEyeHeight;
 
    public Entity(EntityType<?> entityTypeIn, World worldIn) {
+      super(Entity.class);
       this.type = entityTypeIn;
       this.world = worldIn;
       this.dimensions = entityTypeIn.getDimensions();
@@ -211,7 +213,9 @@
       this.dataTracker.startTracking(NO_GRAVITY, false);
       this.dataTracker.startTracking(POSE, EntityPose.STANDING);
       this.initDataTracker();
-      this.standingEyeHeight = this.getEyeHeight(EntityPose.STANDING, this.dimensions);
+      this.standingEyeHeight = getEyeHeightForge(EntityPose.STANDING, this.dimensions);
+      net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(new net.minecraftforge.event.entity.EntityEvent.EntityConstructing(this));
+      this.gatherCapabilities();
    }
 
    @Environment(EnvType.CLIENT)
@@ -290,7 +294,7 @@
    @Environment(EnvType.CLIENT)
    protected void afterSpawn() {
       if (this.world != null) {
-         for(double d0 = this.getY(); d0 > 0.0D && d0 < 256.0D; ++d0) {
+         for(double d0 = this.getY(); d0 > 0.0D && d0 < this.world.getDimension().getHeight(); ++d0) {
             this.updatePosition(this.getX(), d0, this.getZ());
             if (this.world.doesNotCollide(this)) {
                break;
@@ -303,7 +307,13 @@
    }
 
    public void remove() {
+      this.remove(false);
+   }
+
+   public void remove(boolean keepData) {
       this.removed = true;
+      if (!keepData)
+         this.invalidateCaps();
    }
 
    protected void setPose(EntityPose poseIn) {
@@ -321,6 +331,7 @@
 
    public void updatePosition(double x, double y, double z) {
       this.setPos(x, y, z);
+      if (this.isAddedToWorld() && !this.world.isClient && world instanceof ServerWorld) ((ServerWorld)this.world).checkChunk(this); // Forge - Process chunk registration after moving.
       float f = this.dimensions.width / 2.0F;
       float f1 = this.dimensions.height;
       this.setBoundingBox(new Box(x - (double)f, y, z - (double)f, x + (double)f, y + (double)f1, z + (double)f));
@@ -521,7 +532,7 @@
 
             this.horizontalSpeed = (float)((double)this.horizontalSpeed + (double)MathHelper.sqrt(squaredHorizontalLength(vec3d)) * 0.6D);
             this.distanceTraveled = (float)((double)this.distanceTraveled + (double)MathHelper.sqrt(d0 * d0 + d1 * d1 + d2 * d2) * 0.6D);
-            if (this.distanceTraveled > this.nextStepSoundDistance && !blockstate.isAir()) {
+            if (this.distanceTraveled > this.nextStepSoundDistance && !blockstate.isAir(this.world, blockpos)) {
                this.nextStepSoundDistance = this.calculateNextStepSoundDistance();
                if (this.isTouchingWater()) {
                   Entity entity = this.hasPassengers() && this.getPrimaryPassenger() != null ? this.getPrimaryPassenger() : this;
@@ -536,7 +547,7 @@
                } else {
                   this.playStepSound(blockpos, blockstate);
                }
-            } else if (this.distanceTraveled > this.nextFlySoundDistance && this.hasWings() && blockstate.isAir()) {
+            } else if (this.distanceTraveled > this.nextFlySoundDistance && this.hasWings() && blockstate.isAir(this.world, blockpos)) {
                this.nextFlySoundDistance = this.playFlySound(this.distanceTraveled);
             }
          }
@@ -580,11 +591,10 @@
       int j = MathHelper.floor(this.y - (double)0.2F);
       int k = MathHelper.floor(this.z);
       BlockPos blockpos = new BlockPos(i, j, k);
-      if (this.world.getBlockState(blockpos).isAir()) {
+      if (this.world.isAir(blockpos)) {
          BlockPos blockpos1 = blockpos.down();
          BlockState blockstate = this.world.getBlockState(blockpos1);
-         Block block = blockstate.getBlock();
-         if (block.matches(BlockTags.FENCES) || block.matches(BlockTags.WALLS) || block instanceof FenceGateBlock) {
+         if (blockstate.collisionExtendsVertically(this.world, blockpos1, this)) {
             return blockpos1;
          }
       }
@@ -768,6 +778,7 @@
    public void moveToBoundingBoxCenter() {
       Box axisalignedbb = this.getBoundingBox();
       this.setPos((axisalignedbb.x1 + axisalignedbb.x2) / 2.0D, axisalignedbb.y1, (axisalignedbb.z1 + axisalignedbb.z2) / 2.0D);
+      if (this.isAddedToWorld() && !this.world.isClient && world instanceof ServerWorld) ((ServerWorld)this.world).checkChunk(this); // Forge - Process chunk registration after moving.
    }
 
    protected SoundEvent getSwimSound() {
@@ -820,7 +831,7 @@
    protected void playStepSound(BlockPos pos, BlockState blockIn) {
       if (!blockIn.getMaterial().isLiquid()) {
          BlockState blockstate = this.world.getBlockState(pos.up());
-         BlockSoundGroup soundtype = blockstate.getBlock() == Blocks.SNOW ? blockstate.getSoundGroup() : blockIn.getSoundGroup();
+         BlockSoundGroup soundtype = blockstate.getBlock() == Blocks.SNOW ? blockstate.getSoundType(world, pos, this) : blockIn.getSoundType(world, pos, this);
          this.playSound(soundtype.getStepSound(), soundtype.getVolume() * 0.15F, soundtype.getPitch());
       }
    }
@@ -1017,9 +1028,10 @@
       int k = MathHelper.floor(this.getZ());
       BlockPos blockpos = new BlockPos(i, j, k);
       BlockState blockstate = this.world.getBlockState(blockpos);
+      if (!blockstate.addRunningEffects(world, blockpos, this))
       if (blockstate.getRenderType() != BlockRenderType.INVISIBLE) {
          Vec3d vec3d = this.getVelocity();
-         this.world.addParticle(new BlockStateParticleEffect(ParticleTypes.BLOCK, blockstate), this.getX() + ((double)this.random.nextFloat() - 0.5D) * (double)this.dimensions.width, this.getY() + 0.1D, this.getZ() + ((double)this.random.nextFloat() - 0.5D) * (double)this.dimensions.width, vec3d.x * -4.0D, 1.5D, vec3d.z * -4.0D);
+         this.world.addParticle(new BlockStateParticleEffect(ParticleTypes.BLOCK, blockstate).setPos(blockpos), this.getX() + ((double)this.random.nextFloat() - 0.5D) * (double)this.dimensions.width, this.getY() + 0.1D, this.getZ() + ((double)this.random.nextFloat() - 0.5D) * (double)this.dimensions.width, vec3d.x * -4.0D, 1.5D, vec3d.z * -4.0D);
       }
 
    }
@@ -1038,7 +1050,7 @@
             return false;
          } else {
             FluidState ifluidstate = this.world.getFluidState(blockpos);
-            return ifluidstate.matches(p_213290_1_) && d0 < (double)((float)blockpos.getY() + ifluidstate.getHeight(this.world, blockpos) + 0.11111111F);
+            return ifluidstate.isEntityInside(world, blockpos, this, d0, p_213290_1_, true);
          }
       }
    }
@@ -1328,6 +1340,7 @@
          if (this.glowing) {
             compound.putBoolean("Glowing", this.glowing);
          }
+         compound.putBoolean("CanUpdate", canUpdate);
 
          if (!this.scoreboardTags.isEmpty()) {
             ListTag listnbt = new ListTag();
@@ -1339,6 +1352,10 @@
             compound.put("Tags", listnbt);
          }
 
+         CompoundTag caps = serializeCaps();
+         if (caps != null) compound.put("ForgeCaps", caps);
+         if (persistentData != null) compound.put("ForgeData", persistentData);
+
          this.writeCustomDataToTag(compound);
          if (this.hasPassengers()) {
             ListTag listnbt1 = new ListTag();
@@ -1407,6 +1424,9 @@
                this.setSilent(compound.getBoolean("Silent"));
                this.setNoGravity(compound.getBoolean("NoGravity"));
                this.setGlowing(compound.getBoolean("Glowing"));
+               if (compound.contains("ForgeData", 10)) persistentData = compound.getCompound("ForgeData");
+               if (compound.contains("CanUpdate", 99)) this.canUpdate(compound.getBoolean("CanUpdate"));
+               if (compound.contains("ForgeCaps", 10)) deserializeCaps(compound.getCompound("ForgeCaps"));
                if (compound.contains("Tags", 9)) {
                   this.scoreboardTags.clear();
                   ListTag listnbt1 = compound.getList("Tags", 8);
@@ -1495,6 +1515,8 @@
       } else {
          ItemEntity itementity = new ItemEntity(this.world, this.getX(), this.getY() + (double)offsetY, this.getZ(), stack);
          itementity.setToDefaultPickupDelay();
+         if (captureDrops() != null) captureDrops().add(itementity);
+         else
          this.world.spawnEntity(itementity);
          return itementity;
       }
@@ -1538,6 +1560,7 @@
 
    public void tickRiding() {
       this.setVelocity(Vec3d.ZERO);
+      if (canUpdate())
       this.tick();
       if (this.hasVehicle()) {
          this.getVehicle().updatePassengerPosition(this);
@@ -1582,6 +1605,7 @@
          }
       }
 
+      if (!net.minecraftforge.event.ForgeEventFactory.canMountEntity(this, entityIn, true)) return false;
       if (force || this.canStartRiding(entityIn) && entityIn.canAddPassenger(this)) {
          if (this.hasVehicle()) {
             this.stopRiding();
@@ -1613,6 +1637,7 @@
    public void stopRiding() {
       if (this.vehicle != null) {
          Entity entity = this.vehicle;
+         if (!net.minecraftforge.event.ForgeEventFactory.canMountEntity(this, entity, false)) return;
          this.vehicle = null;
          entity.removePassenger(this);
       }
@@ -1769,6 +1794,7 @@
       return !this.getPassengerList().isEmpty();
    }
 
+   @Deprecated //Forge: Use rider sensitive version
    public boolean canBeRiddenInWater() {
       return true;
    }
@@ -1987,7 +2013,7 @@
    }
 
    protected Text getDefaultName() {
-      return this.type.getName();
+      return this.getType().getName(); // Forge: Use getter to allow overriding by mods
    }
 
    public boolean isPartOf(Entity entityIn) {
@@ -2044,6 +2070,11 @@
 
    @Nullable
    public Entity changeDimension(DimensionType destination) {
+      return this.changeDimension(destination, getServer().getWorld(destination).getPortalForcer());
+   }
+   @Nullable
+   public Entity changeDimension(DimensionType destination, net.minecraftforge.common.util.ITeleporter teleporter) {
+      if (!net.minecraftforge.common.ForgeHooks.onTravelToDimension(this, destination)) return null;
       if (!this.world.isClient && !this.removed) {
          this.world.getProfiler().push("changeDimension");
          MinecraftServer minecraftserver = this.getServer();
@@ -2053,6 +2084,7 @@
          this.dimension = destination;
          this.detach();
          this.world.getProfiler().push("reposition");
+         Entity transportedEntity = teleporter.placeEntity(this, serverworld, serverworld1, this.yaw, spawnPortal -> { //Forge: Start vanilla logic
          Vec3d vec3d = this.getVelocity();
          float f = 0.0F;
          BlockPos blockpos;
@@ -2061,16 +2093,9 @@
          } else if (destination == DimensionType.THE_END) {
             blockpos = serverworld1.getForcedSpawnPoint();
          } else {
-            double d0 = this.getX();
-            double d1 = this.getZ();
-            double d2 = 8.0D;
-            if (dimensiontype == DimensionType.OVERWORLD && destination == DimensionType.THE_NETHER) {
-               d0 /= 8.0D;
-               d1 /= 8.0D;
-            } else if (dimensiontype == DimensionType.THE_NETHER && destination == DimensionType.OVERWORLD) {
-               d0 *= 8.0D;
-               d1 *= 8.0D;
-            }
+            double movementFactor = serverworld.getDimension().getMovementFactor() / serverworld1.getDimension().getMovementFactor();
+            double d0 = this.getX() * movementFactor;
+            double d1 = this.getZ() * movementFactor;
 
             double d3 = Math.min(-2.9999872E7D, serverworld1.getWorldBorder().getBoundWest() + 16.0D);
             double d4 = Math.min(-2.9999872E7D, serverworld1.getWorldBorder().getBoundNorth() + 16.0D);
@@ -2080,6 +2105,7 @@
             d1 = MathHelper.clamp(d1, d4, d6);
             Vec3d vec3d1 = this.getLastNetherPortalDirectionVector();
             blockpos = new BlockPos(d0, this.getY(), d1);
+            if (spawnPortal) {
             BlockPattern.TeleportTarget blockpattern$portalinfo = serverworld1.getPortalForcer().getPortal(blockpos, vec3d, this.getLastNetherPortalDirection(), vec3d1.x, vec3d1.y, this instanceof PlayerEntity);
             if (blockpattern$portalinfo == null) {
                return null;
@@ -2088,6 +2114,7 @@
             blockpos = new BlockPos(blockpattern$portalinfo.pos);
             vec3d = blockpattern$portalinfo.velocity;
             f = (float)blockpattern$portalinfo.yaw;
+            }
          }
 
          this.world.getProfiler().swap("reloading");
@@ -2098,13 +2125,15 @@
             entity.setVelocity(vec3d);
             serverworld1.onDimensionChanged(entity);
          }
+            return entity;
+         });//Forge: End vanilla logic
 
-         this.removed = true;
+         this.remove(false);
          this.world.getProfiler().pop();
          serverworld.resetIdleTimeout();
          serverworld1.resetIdleTimeout();
          this.world.getProfiler().pop();
-         return entity;
+         return transportedEntity;
       } else {
          return null;
       }
@@ -2127,10 +2156,12 @@
    }
 
    public Vec3d getLastNetherPortalDirectionVector() {
+      if (this.lastNetherPortalDirectionVector == null) return Vec3d.ZERO;
       return this.lastNetherPortalDirectionVector;
    }
 
    public Direction getLastNetherPortalDirection() {
+      if (this.lastNetherPortalDirection == null) return Direction.NORTH;
       return this.lastNetherPortalDirection;
    }
 
@@ -2259,7 +2290,7 @@
       EntityPose pose = this.getPose();
       EntityDimensions entitysize1 = this.getDimensions(pose);
       this.dimensions = entitysize1;
-      this.standingEyeHeight = this.getEyeHeight(pose, entitysize1);
+      this.standingEyeHeight = getEyeHeightForge(pose, entitysize1);
       if (entitysize1.width < entitysize.width) {
          double d0 = (double)entitysize1.width / 2.0D;
          this.setBoundingBox(new Box(this.getX() - d0, this.getY(), this.getZ() - d0, this.getX() + d0, this.getY() + (double)entitysize1.height, this.getZ() + d0));
@@ -2711,6 +2742,7 @@
       this.x = x;
       this.y = y;
       this.z = z;
+      if (this.isAddedToWorld() && !this.world.isClient && !this.removed) this.world.getChunk((int) Math.floor(this.x) >> 4, (int) Math.floor(this.z) >> 4); // Forge - ensure target chunk is loaded.
    }
 
    public void checkDespawn() {
@@ -2724,4 +2756,69 @@
    public interface PositionUpdater {
       void accept(Entity p_accept_1_, double p_accept_2_, double p_accept_4_, double p_accept_6_);
    }
+
+   /* ================================== Forge Start =====================================*/
+
+   private boolean canUpdate = true;
+   @Override
+   public void canUpdate(boolean value) {
+      this.canUpdate = value;
+   }
+   @Override
+   public boolean canUpdate() {
+      return this.canUpdate;
+   }
+   private Collection<ItemEntity> captureDrops = null;
+   @Override
+   public Collection<ItemEntity> captureDrops() {
+      return captureDrops;
+   }
+   @Override
+   public Collection<ItemEntity> captureDrops(Collection<ItemEntity> value) {
+      Collection<ItemEntity> ret = captureDrops;
+      this.captureDrops = value;
+      return ret;
+   }
+   private CompoundTag persistentData;
+   @Override
+   public CompoundTag getPersistentData() {
+      if (persistentData == null)
+         persistentData = new CompoundTag();
+      return persistentData;
+   }
+   @Override
+   public boolean canTrample(BlockState state, BlockPos pos, float fallDistance) {
+      return world.random.nextFloat() < fallDistance - 0.5F
+              && this instanceof LivingEntity
+              && (this instanceof PlayerEntity || net.minecraftforge.event.ForgeEventFactory.getMobGriefingEvent(world, this))
+              && this.getWidth() * this.getWidth() * this.getHeight() > 0.512F;
+   }
+
+   /**
+    * Internal use for keeping track of entities that are tracked by a world, to
+    * allow guarantees that entity position changes will force a chunk load, avoiding
+    * potential issues with entity desyncing and bad chunk data.
+    */
+   private boolean isAddedToWorld;
+
+   @Override
+   public final boolean isAddedToWorld() { return this.isAddedToWorld; }
+
+   @Override
+   public void onAddedToWorld() { this.isAddedToWorld = true; }
+
+   @Override
+   public void onRemovedFromWorld() { this.isAddedToWorld = false; }
+
+   @Override
+   public void revive() {
+      this.removed = false;
+      this.reviveCaps();
+   }
+
+   private float getEyeHeightForge(EntityPose pose, EntityDimensions size) {
+      net.minecraftforge.event.entity.EntityEvent.EyeHeight evt = new net.minecraftforge.event.entity.EntityEvent.EyeHeight(this, pose, size, this.getEyeHeight(pose, size));
+      net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(evt);
+      return evt.getNewHeight();
+   }
 }
