--- a/net/minecraft/entity/LivingEntity.java
+++ b/net/minecraft/entity/LivingEntity.java
@@ -109,7 +109,12 @@
 
 public abstract class LivingEntity extends Entity {
    private static final UUID ATTR_SPRINTING_SPEED_BOOST_ID = UUID.fromString("662A6B8D-DA3E-4C1C-8813-96EA6097278D");
+   private static final UUID SLOW_FALLING_ID = UUID.fromString("A5B6CF2A-2F7C-31EF-9022-7C3E7D5E6ABA");
    private static final EntityAttributeModifier ATTR_SPRINTING_SPEED_BOOST = (new EntityAttributeModifier(ATTR_SPRINTING_SPEED_BOOST_ID, "Sprinting speed boost", (double)0.3F, EntityAttributeModifier.Operation.MULTIPLY_TOTAL)).setSerialize(false);
+   private static final EntityAttributeModifier SLOW_FALLING = new EntityAttributeModifier(SLOW_FALLING_ID, "Slow falling acceleration reduction", -0.07, EntityAttributeModifier.Operation.ADDITION).setSerialize(false); // Add -0.07 to 0.08 so we get the vanilla default of 0.01
+   public static final net.minecraft.entity.attribute.EntityAttribute SWIM_SPEED = new net.minecraft.entity.attribute.ClampedEntityAttribute(null, "forge.swimSpeed", 1.0D, 0.0D, 1024.0D).setTracked(true);
+   public static final net.minecraft.entity.attribute.EntityAttribute NAMETAG_DISTANCE = new net.minecraft.entity.attribute.ClampedEntityAttribute(null, "forge.nameTagDistance", 64.0D, 0.0D, Float.MAX_VALUE).setTracked(true);
+   public static final net.minecraft.entity.attribute.EntityAttribute ENTITY_GRAVITY = new net.minecraft.entity.attribute.ClampedEntityAttribute(null, "forge.entity_gravity", 0.08D, -8.0D, 8.0D).setTracked(true);
    protected static final TrackedData<Byte> LIVING_FLAGS = DataTracker.registerData(LivingEntity.class, TrackedDataHandlerRegistry.BYTE);
    private static final TrackedData<Float> HEALTH = DataTracker.registerData(LivingEntity.class, TrackedDataHandlerRegistry.FLOAT);
    private static final TrackedData<Integer> POTION_SWIRLS_COLOR = DataTracker.registerData(LivingEntity.class, TrackedDataHandlerRegistry.INTEGER);
@@ -235,6 +240,9 @@
       this.getAttributes().register(EntityAttributes.MOVEMENT_SPEED);
       this.getAttributes().register(EntityAttributes.ARMOR);
       this.getAttributes().register(EntityAttributes.ARMOR_TOUGHNESS);
+      this.getAttributes().register(SWIM_SPEED);
+      this.getAttributes().register(NAMETAG_DISTANCE);
+      this.getAttributes().register(ENTITY_GRAVITY);
    }
 
    protected void fall(double y, boolean onGroundIn, BlockState state, BlockPos pos) {
@@ -244,9 +252,10 @@
 
       if (!this.world.isClient && this.fallDistance > 3.0F && onGroundIn) {
          float f = (float)MathHelper.ceil(this.fallDistance - 3.0F);
-         if (!state.isAir()) {
+         if (!state.isAir(world, pos)) {
             double d0 = Math.min((double)(0.2F + f / 15.0F), 2.5D);
             int i = (int)(150.0D * d0);
+            if (!state.addLandingEffects((ServerWorld)this.world, pos, state, this, i))
             ((ServerWorld)this.world).spawnParticles(new BlockStateParticleEffect(ParticleTypes.BLOCK, state), this.getX(), this.getY(), this.getZ(), i, 0.0D, 0.0D, 0.0D, (double)0.15F);
          }
       }
@@ -310,7 +319,7 @@
                }
             }
 
-            if (!this.world.isClient && this.hasVehicle() && this.getVehicle() != null && !this.getVehicle().canBeRiddenInWater()) {
+            if (!this.world.isClient && this.hasVehicle() && this.getVehicle() != null && !this.getVehicle().canBeRiddenInWater(this)) {
                this.stopRiding();
             }
          } else if (this.getAir() < this.getMaxAir()) {
@@ -392,7 +401,7 @@
    protected void updatePostDeath() {
       ++this.deathTime;
       if (this.deathTime == 20) {
-         this.remove();
+         this.remove(this instanceof net.minecraft.server.network.ServerPlayerEntity); //Forge keep data until we revive player
 
          for(int i = 0; i < 20; ++i) {
             double d0 = this.random.nextGaussian() * 0.02D;
@@ -573,7 +582,7 @@
             if (!effectinstance.update(this, () -> {
                this.onStatusEffectUpgraded(effectinstance, true);
             })) {
-               if (!this.world.isClient) {
+               if (!this.world.isClient && !net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(new net.minecraftforge.event.entity.living.PotionEvent.PotionExpiryEvent(this, effectinstance))) {
                   iterator.remove();
                   this.onStatusEffectRemoved(effectinstance);
                }
@@ -623,8 +632,10 @@
          this.setInvisible(false);
       } else {
          Collection<StatusEffectInstance> collection = this.activeStatusEffects.values();
-         this.dataTracker.set(POTION_SWIRLS_AMBIENT, containsOnlyAmbientEffects(collection));
-         this.dataTracker.set(POTION_SWIRLS_COLOR, PotionUtil.getColor(collection));
+         net.minecraftforge.event.entity.living.PotionColorCalculationEvent event = new net.minecraftforge.event.entity.living.PotionColorCalculationEvent(this, PotionUtil.getColor(collection), containsOnlyAmbientEffects(collection), collection);
+         net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(event);
+         this.dataTracker.set(POTION_SWIRLS_AMBIENT, event.areParticlesHidden());
+         this.dataTracker.set(POTION_SWIRLS_COLOR, event.getColor());
          this.setInvisible(this.hasStatusEffect(StatusEffects.INVISIBILITY));
       }
 
@@ -688,7 +699,9 @@
 
          boolean flag;
          for(flag = false; iterator.hasNext(); flag = true) {
-            this.onStatusEffectRemoved(iterator.next());
+            StatusEffectInstance effect = iterator.next();
+            if(net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(new net.minecraftforge.event.entity.living.PotionEvent.PotionRemoveEvent(this, effect))) continue;
+            this.onStatusEffectRemoved(effect);
             iterator.remove();
          }
 
@@ -718,6 +731,7 @@
          return false;
       } else {
          StatusEffectInstance effectinstance = this.activeStatusEffects.get(effectInstanceIn.getEffectType());
+         net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(new net.minecraftforge.event.entity.living.PotionEvent.PotionAddedEvent(this, effectinstance, effectInstanceIn));
          if (effectinstance == null) {
             this.activeStatusEffects.put(effectInstanceIn.getEffectType(), effectInstanceIn);
             this.onStatusEffectApplied(effectInstanceIn);
@@ -732,6 +746,9 @@
    }
 
    public boolean canHaveStatusEffect(StatusEffectInstance potioneffectIn) {
+      net.minecraftforge.event.entity.living.PotionEvent.PotionApplicableEvent event = new net.minecraftforge.event.entity.living.PotionEvent.PotionApplicableEvent(this, potioneffectIn);
+      net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(event);
+      if (event.getResult() != net.minecraftforge.eventbus.api.Event.Result.DEFAULT) return event.getResult() == net.minecraftforge.eventbus.api.Event.Result.ALLOW;
       if (this.getGroup() == EntityGroup.UNDEAD) {
          StatusEffect effect = potioneffectIn.getEffectType();
          if (effect == StatusEffects.REGENERATION || effect == StatusEffects.POISON) {
@@ -752,6 +769,7 @@
    }
 
    public boolean removeStatusEffect(StatusEffect effectIn) {
+      if (net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(new net.minecraftforge.event.entity.living.PotionEvent.PotionRemoveEvent(this, effectIn))) return false;
       StatusEffectInstance effectinstance = this.removeStatusEffectInternal(effectIn);
       if (effectinstance != null) {
          this.onStatusEffectRemoved(effectinstance);
@@ -788,6 +806,8 @@
    }
 
    public void heal(float healAmount) {
+      healAmount = net.minecraftforge.event.ForgeEventFactory.onLivingHeal(this, healAmount);
+      if (healAmount <= 0) return;
       float f = this.getHealth();
       if (f > 0.0F) {
          this.setHealth(f + healAmount);
@@ -804,6 +824,7 @@
    }
 
    public boolean damage(DamageSource source, float amount) {
+      if (!net.minecraftforge.common.ForgeHooks.onLivingAttack(this, source, amount)) return false;
       if (this.isInvulnerableTo(source)) {
          return false;
       } else if (this.world.isClient) {
@@ -870,8 +891,8 @@
             if (entity1 instanceof PlayerEntity) {
                this.playerHitTimer = 100;
                this.attackingPlayer = (PlayerEntity)entity1;
-            } else if (entity1 instanceof WolfEntity) {
-               WolfEntity wolfentity = (WolfEntity)entity1;
+            } else if (entity1 instanceof net.minecraft.entity.passive.TameableEntity) {
+               net.minecraft.entity.passive.TameableEntity wolfentity = (net.minecraft.entity.passive.TameableEntity)entity1;
                if (wolfentity.isTamed()) {
                   this.playerHitTimer = 100;
                   LivingEntity livingentity = wolfentity.getOwner();
@@ -1053,6 +1074,7 @@
    }
 
    public void onDeath(DamageSource cause) {
+      if (net.minecraftforge.common.ForgeHooks.onLivingDeath(this, cause)) return;
       if (!this.removed && !this.dead) {
          Entity entity = cause.getAttacker();
          LivingEntity livingentity = this.getPrimeAdversary();
@@ -1084,10 +1106,10 @@
       if (!this.world.isClient) {
          boolean flag = false;
          if (p_226298_1_ instanceof WitherEntity) {
-            if (this.world.getGameRules().getBoolean(GameRules.MOB_GRIEFING)) {
+               if (net.minecraftforge.event.ForgeEventFactory.getMobGriefingEvent(this.world, this)) {
                BlockPos blockpos = new BlockPos(this);
                BlockState blockstate = Blocks.WITHER_ROSE.getDefaultState();
-               if (this.world.getBlockState(blockpos).isAir() && blockstate.canPlaceAt(this.world, blockpos)) {
+               if (this.world.isAir(blockpos) && blockstate.canPlaceAt(this.world, blockpos)) {
                   this.world.setBlockState(blockpos, blockstate, 3);
                   flag = true;
                }
@@ -1104,13 +1126,10 @@
 
    protected void drop(DamageSource damageSourceIn) {
       Entity entity = damageSourceIn.getAttacker();
-      int i;
-      if (entity instanceof PlayerEntity) {
-         i = EnchantmentHelper.getLooting((LivingEntity)entity);
-      } else {
-         i = 0;
-      }
 
+      int i = net.minecraftforge.common.ForgeHooks.getLootingLevel(this, entity, damageSourceIn);
+      this.captureDrops(new java.util.ArrayList<>());
+
       boolean flag = this.playerHitTimer > 0;
       if (this.canDropLootAndXp() && this.world.getGameRules().getBoolean(GameRules.DO_MOB_LOOT)) {
          this.dropLoot(damageSourceIn, flag);
@@ -1119,6 +1138,10 @@
 
       this.dropInventory();
       this.dropXp();
+
+      Collection<ItemEntity> drops = captureDrops(null);
+      if (!net.minecraftforge.common.ForgeHooks.onLivingDrops(this, damageSourceIn, drops, i, playerHitTimer > 0))
+         drops.forEach(e -> world.spawnEntity(e));
    }
 
    protected void dropInventory() {
@@ -1128,6 +1151,7 @@
       if (!this.world.isClient && (this.shouldAlwaysDropXp() || this.playerHitTimer > 0 && this.canDropLootAndXp() && this.world.getGameRules().getBoolean(GameRules.DO_MOB_LOOT))) {
          int i = this.getCurrentExperience(this.attackingPlayer);
 
+         i = net.minecraftforge.event.ForgeEventFactory.getExperienceDrop(this, this.attackingPlayer, i);
          while(i > 0) {
             int j = ExperienceOrbEntity.roundToOrbSize(i);
             i -= j;
@@ -1135,6 +1159,7 @@
          }
       }
 
+
    }
 
    protected void dropEquipment(DamageSource source, int looting, boolean recentlyHitIn) {
@@ -1148,7 +1173,8 @@
       Identifier resourcelocation = this.getLootTable();
       LootTable loottable = this.world.getServer().getLootManager().getSupplier(resourcelocation);
       LootContext.Builder lootcontext$builder = this.getLootContextBuilder(p_213354_2_, damageSourceIn);
-      loottable.dropLimited(lootcontext$builder.build(LootContextTypes.ENTITY), this::dropStack);
+      LootContext ctx = lootcontext$builder.build(LootContextTypes.ENTITY);
+      loottable.getDrops(ctx).forEach(this::dropStack);
    }
 
    protected LootContext.Builder getLootContextBuilder(boolean p_213363_1_, DamageSource damageSourceIn) {
@@ -1161,6 +1187,9 @@
    }
 
    public void takeKnockback(Entity entityIn, float strength, double xRatio, double zRatio) {
+      net.minecraftforge.event.entity.living.LivingKnockBackEvent event = net.minecraftforge.common.ForgeHooks.onLivingKnockBack(this, entityIn, strength, xRatio, zRatio);
+      if(event.isCanceled()) return;
+      strength = event.getStrength(); xRatio = event.getRatioX(); zRatio = event.getRatioZ();
       if (!(this.random.nextDouble() < this.getAttributeInstance(EntityAttributes.KNOCKBACK_RESISTANCE).getValue())) {
          this.velocityDirty = true;
          Vec3d vec3d = this.getVelocity();
@@ -1196,12 +1225,7 @@
          return false;
       } else {
          BlockState blockstate = this.getBlockState();
-         Block block = blockstate.getBlock();
-         if (block != Blocks.LADDER && block != Blocks.VINE && block != Blocks.SCAFFOLDING) {
-            return block instanceof TrapdoorBlock && this.canEnterTrapdoor(new BlockPos(this), blockstate);
-         } else {
-            return true;
-         }
+         return net.minecraftforge.common.ForgeHooks.isLivingOnLadder(blockstate, world, new BlockPos(this), this);
       }
    }
 
@@ -1225,6 +1249,11 @@
    }
 
    public boolean handleFallDamage(float distance, float damageMultiplier) {
+      float[] ret = net.minecraftforge.common.ForgeHooks.onLivingFall(this, distance, damageMultiplier);
+      if (ret == null) return false;
+      distance = ret[0];
+      damageMultiplier = ret[1];
+
       boolean flag = super.handleFallDamage(distance, damageMultiplier);
       int i = this.computeFallDamage(distance, damageMultiplier);
       if (i > 0) {
@@ -1248,9 +1277,10 @@
          int i = MathHelper.floor(this.getX());
          int j = MathHelper.floor(this.getY() - (double)0.2F);
          int k = MathHelper.floor(this.getZ());
-         BlockState blockstate = this.world.getBlockState(new BlockPos(i, j, k));
-         if (!blockstate.isAir()) {
-            BlockSoundGroup soundtype = blockstate.getSoundGroup();
+         BlockPos pos = new BlockPos(i, j, k);
+         BlockState blockstate = this.world.getBlockState(pos);
+         if (!blockstate.isAir(this.world, pos)) {
+            BlockSoundGroup soundtype = blockstate.getSoundType(world, pos, this);
             this.playSound(soundtype.getFallSound(), soundtype.getVolume() * 0.5F, soundtype.getPitch() * 0.75F);
          }
 
@@ -1319,6 +1349,8 @@
 
    protected void applyDamage(DamageSource damageSrc, float damageAmount) {
       if (!this.isInvulnerableTo(damageSrc)) {
+         damageAmount = net.minecraftforge.common.ForgeHooks.onLivingHurt(this, damageSrc, damageAmount);
+         if (damageAmount <= 0) return;
          damageAmount = this.applyArmorToDamage(damageSrc, damageAmount);
          damageAmount = this.applyEnchantmentsToDamage(damageSrc, damageAmount);
          float f2 = Math.max(damageAmount - this.getAbsorptionAmount(), 0.0F);
@@ -1328,10 +1360,11 @@
             ((ServerPlayerEntity)damageSrc.getAttacker()).increaseStat(Stats.DAMAGE_DEALT_ABSORBED, Math.round(f * 10.0F));
          }
 
+         f2 = net.minecraftforge.common.ForgeHooks.onLivingDamage(this, damageSrc, f2);
          if (f2 != 0.0F) {
             float f1 = this.getHealth();
-            this.setHealth(f1 - f2);
             this.getDamageTracker().onDamage(damageSrc, f1, f2);
+            this.setHealth(f1 - f2); // Forge: moved to fix MC-121048
             this.setAbsorptionAmount(this.getAbsorptionAmount() - f2);
          }
       }
@@ -1385,6 +1418,8 @@
    }
 
    public void swingHand(Hand handIn, boolean p_226292_2_) {
+      ItemStack stack = this.getStackInHand(handIn);
+      if (!stack.isEmpty() && stack.onEntitySwing(this)) return;
       if (!this.isHandSwinging || this.handSwingTicks >= this.getHandSwingDuration() / 2 || this.handSwingTicks < 0) {
          this.handSwingTicks = -1;
          this.isHandSwinging = true;
@@ -1771,15 +1806,16 @@
       }
 
       this.velocityDirty = true;
+      net.minecraftforge.common.ForgeHooks.onLivingJump(this);
    }
 
    @Environment(EnvType.CLIENT)
    protected void knockDownwards() {
-      this.setVelocity(this.getVelocity().add(0.0D, (double)-0.04F, 0.0D));
+      this.setVelocity(this.getVelocity().add(0.0D, (double)-0.04F  * this.getAttributeInstance(SWIM_SPEED).getValue(), 0.0D));
    }
 
    protected void swimUpward(Tag<Fluid> fluidTag) {
-      this.setVelocity(this.getVelocity().add(0.0D, (double)0.04F, 0.0D));
+      this.setVelocity(this.getVelocity().add(0.0D, (double)0.04F * this.getAttributeInstance(SWIM_SPEED).getValue(), 0.0D));
    }
 
    protected float getBaseMovementSpeedMultiplier() {
@@ -1789,11 +1825,15 @@
    public void travel(Vec3d p_213352_1_) {
       if (this.canMoveVoluntarily() || this.isLogicalSideForUpdatingMovement()) {
          double d0 = 0.08D;
+         EntityAttributeInstance gravity = this.getAttributeInstance(ENTITY_GRAVITY);
          boolean flag = this.getVelocity().y <= 0.0D;
          if (flag && this.hasStatusEffect(StatusEffects.SLOW_FALLING)) {
-            d0 = 0.01D;
+            if (!gravity.hasModifier(SLOW_FALLING)) gravity.addModifier(SLOW_FALLING);
             this.fallDistance = 0.0F;
+         } else if (gravity.hasModifier(SLOW_FALLING)) {
+            gravity.removeModifier(SLOW_FALLING);
          }
+         d0 = gravity.getValue();
 
          if (!this.isTouchingWater() || this instanceof PlayerEntity && ((PlayerEntity)this).abilities.flying) {
             if (!this.isInLava() || this instanceof PlayerEntity && ((PlayerEntity)this).abilities.flying) {
@@ -1842,7 +1882,7 @@
                   }
                } else {
                   BlockPos blockpos = this.getVelocityAffectingPos();
-                  float f5 = this.world.getBlockState(blockpos).getBlock().getSlipperiness();
+                  float f5 = this.world.getBlockState(blockpos).getSlipperiness(world, blockpos, this);
                   float f7 = this.onGround ? f5 * 0.91F : 0.91F;
                   this.updateVelocity(this.getMovementSpeed(f5), p_213352_1_);
                   this.setVelocity(this.applyClimbingSpeed(this.getVelocity()));
@@ -1904,6 +1944,7 @@
                f = 0.96F;
             }
 
+            f1 *= (float)this.getAttributeInstance(SWIM_SPEED).getValue();
             this.updateVelocity(f1, p_213352_1_);
             this.move(MovementType.SELF, this.getVelocity());
             Vec3d vec3d1 = this.getVelocity();
@@ -1979,6 +2020,7 @@
    }
 
    public void tick() {
+      if (net.minecraftforge.common.ForgeHooks.onLivingUpdate(this)) return;
       super.tick();
       this.tickActiveItemStack();
       this.updateLeaningPitch();
@@ -2022,7 +2064,9 @@
 
             ItemStack itemstack1 = this.getEquippedStack(equipmentslottype);
             if (!ItemStack.areEqualIgnoreDamage(itemstack1, itemstack)) {
+               if (!itemstack1.equals(itemstack, true))
                ((ServerWorld)this.world).getChunkManager().sendToOtherNearbyPlayers(this, new EntityEquipmentUpdateS2CPacket(this.getEntityId(), equipmentslottype, itemstack1));
+               net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(new net.minecraftforge.event.entity.living.LivingEquipmentChangeEvent(this, equipmentslottype, itemstack, itemstack1));
                if (!itemstack.isEmpty()) {
                   this.getAttributes().removeAll(itemstack.getAttributeModifiers(equipmentslottype));
                }
@@ -2474,13 +2518,22 @@
 
    private void tickActiveItemStack() {
       if (this.isUsingItem()) {
-         if (ItemStack.areItemsEqual(this.getStackInHand(this.getActiveHand()), this.activeItemStack)) {
+         ItemStack itemstack = this.getStackInHand(this.getActiveHand());
+         if (net.minecraftforge.common.ForgeHooks.canContinueUsing(this.activeItemStack, itemstack)) this.activeItemStack = itemstack;
+         if (itemstack == this.activeItemStack) {
+
+            if (!this.activeItemStack.isEmpty()) {
+               itemUseTimeLeft = net.minecraftforge.event.ForgeEventFactory.onItemUseTick(this, activeItemStack, itemUseTimeLeft);
+               if (itemUseTimeLeft > 0)
+                  activeItemStack.onUsingTick(this, itemUseTimeLeft);
+            }
+
             this.activeItemStack.usageTick(this.world, this, this.getItemUseTimeLeft());
             if (this.shouldSpawnConsumptionEffects()) {
                this.spawnConsumptionEffects(this.activeItemStack, 5);
             }
 
-            if (--this.itemUseTimeLeft == 0 && !this.world.isClient && !this.activeItemStack.isUsedOnRelease()) {
+            if (--this.itemUseTimeLeft <= 0 && !this.world.isClient && !this.activeItemStack.isUsedOnRelease()) {
                this.consumeItem();
             }
          } else {
@@ -2522,8 +2575,10 @@
    public void setCurrentHand(Hand hand) {
       ItemStack itemstack = this.getStackInHand(hand);
       if (!itemstack.isEmpty() && !this.isUsingItem()) {
+         int duration = net.minecraftforge.event.ForgeEventFactory.onItemUseStart(this, itemstack, itemstack.getMaxUseTime());
+         if (duration <= 0) return;
          this.activeItemStack = itemstack;
-         this.itemUseTimeLeft = itemstack.getMaxUseTime();
+         this.itemUseTimeLeft = duration;
          if (!this.world.isClient) {
             this.setLivingFlag(1, true);
             this.setLivingFlag(2, hand == Hand.OFF_HAND);
@@ -2583,6 +2638,9 @@
          vec3d1 = vec3d1.rotateX(-this.pitch * ((float)Math.PI / 180F));
          vec3d1 = vec3d1.rotateY(-this.yaw * ((float)Math.PI / 180F));
          vec3d1 = vec3d1.add(this.getX(), this.getEyeY(), this.getZ());
+         if (this.world instanceof ServerWorld) //Forge: Fix MC-2518 spawnParticle is nooped on server, need to use server specific variant
+            ((ServerWorld)this.world).spawnParticles(new ItemStackParticleEffect(ParticleTypes.ITEM, stack), vec3d1.x, vec3d1.y, vec3d1.z, 1, vec3d.x, vec3d.y + 0.05D, vec3d.z, 0.0D);
+         else
          this.world.addParticle(new ItemStackParticleEffect(ParticleTypes.ITEM, stack), vec3d1.x, vec3d1.y, vec3d1.z, vec3d.x, vec3d.y + 0.05D, vec3d.z);
       }
 
@@ -2594,7 +2652,9 @@
       } else {
          if (!this.activeItemStack.isEmpty() && this.isUsingItem()) {
             this.spawnConsumptionEffects(this.activeItemStack, 16);
-            this.setStackInHand(this.getActiveHand(), this.activeItemStack.finishUsing(this.world, this));
+            ItemStack copy = this.activeItemStack.copy();
+            ItemStack stack = net.minecraftforge.event.ForgeEventFactory.onItemUseFinish(this, copy, getItemUseTimeLeft(), this.activeItemStack.finishUsing(this.world, this));
+            this.setStackInHand(this.getActiveHand(), stack);
             this.clearActiveItem();
          }
 
@@ -2615,7 +2675,11 @@
 
    public void stopUsingItem() {
       if (!this.activeItemStack.isEmpty()) {
+         if (!net.minecraftforge.event.ForgeEventFactory.onUseItemStop(this, activeItemStack, this.getItemUseTimeLeft())) {
+            ItemStack copy = this instanceof PlayerEntity ? activeItemStack.copy() : null;
          this.activeItemStack.onStoppedUsing(this.world, this, this.getItemUseTimeLeft());
+           if (copy != null && activeItemStack.isEmpty()) net.minecraftforge.event.ForgeEventFactory.onPlayerDestroyItem((PlayerEntity)this, copy, getActiveHand());
+         }
          if (this.activeItemStack.isUsedOnRelease()) {
             this.tickActiveItemStack();
          }
@@ -2772,16 +2836,16 @@
 
    private boolean isSleepingInBed() {
       return this.getSleepingPosition().map((p_213347_1_) -> {
-         return this.world.getBlockState(p_213347_1_).getBlock() instanceof BedBlock;
+         return net.minecraftforge.event.ForgeEventFactory.fireSleepingLocationCheck(this, p_213347_1_);
       }).orElse(false);
    }
 
    public void wakeUp() {
       this.getSleepingPosition().filter(this.world::isChunkLoaded).ifPresent((p_213368_1_) -> {
          BlockState blockstate = this.world.getBlockState(p_213368_1_);
-         if (blockstate.getBlock() instanceof BedBlock) {
-            this.world.setBlockState(p_213368_1_, blockstate.with(BedBlock.OCCUPIED, Boolean.valueOf(false)), 3);
-            Vec3d vec3d = BedBlock.findWakeUpPosition(this.getType(), this.world, p_213368_1_, 0).orElseGet(() -> {
+         if (blockstate.isBed(this.world, p_213368_1_, this)) {
+            blockstate.setBedOccupied(world, p_213368_1_, this, false);
+            Vec3d vec3d = blockstate.getBedSpawnPosition(this.getType(), world, p_213368_1_, this).orElseGet(()-> {
                BlockPos blockpos = p_213368_1_.up();
                return new Vec3d((double)blockpos.getX() + 0.5D, (double)blockpos.getY() + 0.1D, (double)blockpos.getZ() + 0.5D);
             });
@@ -2797,7 +2861,9 @@
    @Environment(EnvType.CLIENT)
    public Direction getSleepingDirection() {
       BlockPos blockpos = this.getSleepingPosition().orElse((BlockPos)null);
-      return blockpos != null ? BedBlock.getDirection(this.world, blockpos) : null;
+      if (blockpos == null) return Direction.UP;
+      BlockState state = this.world.getBlockState(blockpos);
+      return !state.isBed(world, blockpos, this) ? Direction.UP : state.getBedDirection(world, blockpos);
    }
 
    public boolean isInsideWall() {
@@ -2866,4 +2932,58 @@
    public void sendToolBreakStatus(Hand p_213334_1_) {
       this.sendEquipmentBreakStatus(p_213334_1_ == Hand.MAIN_HAND ? EquipmentSlot.MAINHAND : EquipmentSlot.OFFHAND);
    }
+
+   /* ==== FORGE START ==== */
+   /***
+    * Removes all potion effects that have curativeItem as a curative item for its effect
+    * @param curativeItem The itemstack we are using to cure potion effects
+    */
+   public boolean curePotionEffects(ItemStack curativeItem) {
+      if (this.world.isClient)
+         return false;
+      boolean ret = false;
+      Iterator<StatusEffectInstance> itr = this.activeStatusEffects.values().iterator();
+      while (itr.hasNext()) {
+         StatusEffectInstance effect = itr.next();
+         if (effect.isCurativeItem(curativeItem) && !net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(new net.minecraftforge.event.entity.living.PotionEvent.PotionRemoveEvent(this, effect))) {
+            this.onStatusEffectRemoved(effect);
+            itr.remove();
+            ret = true;
+            this.effectsChanged = true;
+         }
+      }
+      return ret;
+   }
+
+   /**
+    * Returns true if the entity's rider (EntityPlayer) should face forward when mounted.
+    * currently only used in vanilla code by pigs.
+    *
+    * @param player The player who is riding the entity.
+    * @return If the player should orient the same direction as this entity.
+    */
+   public boolean shouldRiderFaceForward(PlayerEntity player) {
+      return this instanceof net.minecraft.entity.passive.PigEntity;
+   }
+
+   private final net.minecraftforge.common.util.LazyOptional<?>[] handlers = net.minecraftforge.items.wrapper.EntityEquipmentInvWrapper.create(this);
+
+   @Override
+   public <T> net.minecraftforge.common.util.LazyOptional<T> getCapability(net.minecraftforge.common.capabilities.Capability<T> capability, @Nullable Direction facing) {
+      if (this.isAlive() && capability == net.minecraftforge.items.CapabilityItemHandler.ITEM_HANDLER_CAPABILITY) {
+         if (facing == null) return handlers[2].cast();
+         else if (facing.getAxis().isVertical()) return handlers[0].cast();
+         else if (facing.getAxis().isHorizontal()) return handlers[1].cast();
+      }
+      return super.getCapability(capability, facing);
+   }
+
+   @Override
+   public void remove(boolean keepData) {
+      super.remove(keepData);
+      if (!keepData) {
+         for (int x = 0; x < handlers.length; x++)
+            handlers[x].invalidate();
+      }
+   }
 }
