--- a/net/minecraft/client/texture/TextureManager.java
+++ b/net/minecraft/client/texture/TextureManager.java
@@ -147,6 +147,7 @@
    public void destroyTexture(Identifier textureLocation) {
       AbstractTexture texture = this.getTexture(textureLocation);
       if (texture != null) {
+         this.textures.remove(textureLocation); // Forge: fix MC-98707
          TextureUtil.releaseTextureId(texture.getGlId());
       }
 
