--- a/net/minecraft/client/network/ServerInfo.java
+++ b/net/minecraft/client/network/ServerInfo.java
@@ -24,6 +24,7 @@
    private ServerInfo.ResourcePackState resourcePackState = ServerInfo.ResourcePackState.PROMPT;
    private String icon;
    private boolean local;
+   public net.minecraftforge.fml.client.ExtendedServerListData forgeData = null;
 
    public ServerInfo(String name, String ip, boolean isLan) {
       this.name = name;
