--- a/net/minecraft/entity/passive/MooshroomEntity.java
+++ b/net/minecraft/entity/passive/MooshroomEntity.java
@@ -34,7 +34,7 @@
 import net.minecraftforge.api.distmarker.OnlyIn;
 import org.apache.commons.lang3.tuple.Pair;
 
-public class MooshroomEntity extends CowEntity {
+public class MooshroomEntity extends CowEntity implements net.minecraftforge.common.IShearable {
    private static final TrackedData<String> TYPE = DataTracker.registerData(MooshroomEntity.class, TrackedDataHandlerRegistry.STRING);
    private StatusEffect stewEffect;
    private int stewEffectDuration;
@@ -98,7 +98,7 @@
 
          this.playSound(soundevent, 1.0F, 1.0F);
          return true;
-      } else if (itemstack.getItem() == Items.SHEARS && !this.isBaby()) {
+      } else if (false && itemstack.getItem() == Items.SHEARS && !this.isBaby()) { //Forge: Moved to onSheared
          this.world.addParticle(ParticleTypes.EXPLOSION, this.getX(), this.getBodyY(0.5D), this.getZ(), 0.0D, 0.0D, 0.0D);
          if (!this.world.isClient) {
             this.remove();
@@ -210,6 +210,34 @@
       return mooshroomentity$type2;
    }
 
+   @Override
+   public boolean isShearable(ItemStack item, net.minecraft.world.WorldView world, net.minecraft.util.math.BlockPos pos) {
+      return !this.isBaby();
+   }
+
+   @Override
+   public java.util.List<ItemStack> onSheared(ItemStack item, net.minecraft.world.IWorld world, net.minecraft.util.math.BlockPos pos, int fortune) {
+      java.util.List<ItemStack> ret = new java.util.ArrayList<>();
+      this.world.addParticle(ParticleTypes.EXPLOSION, this.getX(), this.getBodyY(0.5D), this.getZ(), 0.0D, 0.0D, 0.0D);
+      if (!this.world.isClient) {
+         this.remove();
+         CowEntity cowentity = EntityType.COW.create(this.world);
+         cowentity.refreshPositionAndAngles(this.getX(), this.getY(), this.getZ(), this.yaw, this.pitch);
+         cowentity.setHealth(this.getHealth());
+         cowentity.bodyYaw = this.bodyYaw;
+         if (this.hasCustomName()) {
+             cowentity.setCustomName(this.getCustomName());
+             cowentity.setCustomNameVisible(this.isCustomNameVisible());
+         }
+         this.world.spawnEntity(cowentity);
+         for(int i = 0; i < 5; ++i) {
+            ret.add(new ItemStack(this.getMooshroomType().mushroom.getBlock()));
+         }
+         this.playSound(SoundEvents.ENTITY_MOOSHROOM_SHEAR, 1.0F, 1.0F);
+      }
+      return ret;
+   }
+
    public static enum Type {
       RED("red", Blocks.RED_MUSHROOM.getDefaultState()),
       BROWN("brown", Blocks.BROWN_MUSHROOM.getDefaultState());
