--- a/net/minecraft/entity/passive/RabbitEntity.java
+++ b/net/minecraft/entity/passive/RabbitEntity.java
@@ -492,7 +492,7 @@
 
       public boolean canStart() {
          if (this.cooldown <= 0) {
-            if (!this.rabbit.world.getGameRules().getBoolean(GameRules.MOB_GRIEFING)) {
+            if (!net.minecraftforge.event.ForgeEventFactory.getMobGriefingEvent(this.rabbit.world, this.rabbit)) {
                return false;
             }
 
