--- a/net/minecraft/entity/passive/SnowGolemEntity.java
+++ b/net/minecraft/entity/passive/SnowGolemEntity.java
@@ -33,7 +33,7 @@
 import net.minecraft.world.GameRules;
 import net.minecraft.world.World;
 
-public class SnowGolemEntity extends GolemEntity implements RangedAttackMob {
+public class SnowGolemEntity extends GolemEntity implements RangedAttackMob, net.minecraftforge.common.IShearable {
    private static final TrackedData<Byte> SNOW_GOLEM_FLAGS = DataTracker.registerData(SnowGolemEntity.class, TrackedDataHandlerRegistry.BYTE);
 
    public SnowGolemEntity(EntityType<? extends SnowGolemEntity> type, World worldIn) {
@@ -88,7 +88,7 @@
             this.damage(DamageSource.ON_FIRE, 1.0F);
          }
 
-         if (!this.world.getGameRules().getBoolean(GameRules.MOB_GRIEFING)) {
+         if (!net.minecraftforge.event.ForgeEventFactory.getMobGriefingEvent(this.world, this)) {
             return;
          }
 
@@ -99,7 +99,7 @@
             j = MathHelper.floor(this.getY());
             k = MathHelper.floor(this.getZ() + (double)((float)(l / 2 % 2 * 2 - 1) * 0.25F));
             BlockPos blockpos = new BlockPos(i, j, k);
-            if (this.world.getBlockState(blockpos).isAir() && this.world.getBiome(blockpos).getTemperature(blockpos) < 0.8F && blockstate.canPlaceAt(this.world, blockpos)) {
+            if (this.world.isAir(blockpos) && this.world.getBiome(blockpos).getTemperature(blockpos) < 0.8F && blockstate.canPlaceAt(this.world, blockpos)) {
                this.world.setBlockState(blockpos, blockstate);
             }
          }
@@ -125,7 +125,7 @@
 
    protected boolean interactMob(PlayerEntity player, Hand hand) {
       ItemStack itemstack = player.getStackInHand(hand);
-      if (itemstack.getItem() == Items.SHEARS && this.hasPumpkin()) {
+      if (false && itemstack.getItem() == Items.SHEARS && this.hasPumpkin()) { //Forge: Moved to onSheared
          if (!this.world.isClient) {
             this.setHasPumpkin(false);
             itemstack.damage(1, player, (p_213622_1_) -> {
@@ -167,4 +167,15 @@
    protected SoundEvent getDeathSound() {
       return SoundEvents.ENTITY_SNOW_GOLEM_DEATH;
    }
+
+   @Override
+   public boolean isShearable(ItemStack item, net.minecraft.world.WorldView world, BlockPos pos) {
+      return this.hasPumpkin();
+   }
+
+   @Override
+   public java.util.List<ItemStack> onSheared(ItemStack item, net.minecraft.world.IWorld world, BlockPos pos, int fortune) {
+      this.setHasPumpkin(false);
+      return new java.util.ArrayList<>();
+   }
 }
