--- a/net/minecraft/data/DataCache.java
+++ b/net/minecraft/data/DataCache.java
@@ -23,7 +23,7 @@
    private final Path recordFile;
    private int unchanged;
    private final Map<Path, String> oldSha1 = Maps.newHashMap();
-   private final Map<Path, String> newSha1 = Maps.newHashMap();
+   private final Map<Path, String> newSha1 = Maps.newTreeMap(); //Forge: TreeMap, makes the file output predictable/sorted.
    private final Set<Path> ignores = Sets.newHashSet();
 
    public DataCache(Path folder, String fileName) throws IOException {
@@ -55,7 +55,7 @@
       }
 
       IOUtils.writeLines(this.newSha1.entrySet().stream().map((p_208319_1_) -> {
-         return (String)p_208319_1_.getValue() + ' ' + this.root.relativize(p_208319_1_.getKey());
+         return (String)p_208319_1_.getValue() + ' ' + this.root.relativize(p_208319_1_.getKey()).toString().replace('\\', '/'); //Forge: Standardize file paths.
       }).collect(Collectors.toList()), System.lineSeparator(), writer);
       writer.close();
       LOGGER.debug("Caching: cache hits: {}, created: {} removed: {}", this.unchanged, this.newSha1.size() - this.unchanged, this.oldSha1.size());
